{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use Case 5: Gene set enrichment analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Import standard data analysis imports, as well as the gseapy which will allow us to perform a Gene set enrichment analysis</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy import stats\n",
    "import gseapy as gp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Import the CPTAC data</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading Clinical Data...\n",
      "Loading Proteomics Data...\n",
      "Loading Transcriptomics Data...\n",
      "Loading CNA Data...\n",
      "Loading Phosphoproteomics Data...\n",
      "Loading Somatic Data...\n",
      "\n",
      " ******PLEASE READ******\n",
      "WARNING: This data is under a publication embargo until July 1, 2019. CPTAC is a community resource project and data are made available rapidly after generation for community research use. The embargo allows exploring and utilizing the data, but the data may not be in a publication until July 1, 2019. Please see https://proteomics.cancer.gov/data-portal/about/data-use-agreement or enter CPTAC.embargo() to open the webpage for more details.\n"
     ]
    }
   ],
   "source": [
    "import CPTAC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Retrieve the clinical and proteomics dataframes</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical = CPTAC.get_clinical()\n",
    "proteomics = CPTAC.get_proteomics()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>For this example we will be separating the protein abudance based on the clinical MSI. Our first step is to combine the MSI information into the proteomics dataframe utilizing the <code>CPTAC.compare_clinical()</code> function</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "msiProt = CPTAC.compare_clinical(clinical, proteomics, 'MSI')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Separate the proteomics into two groups based on whether MSI is MSI-H or other </b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "high = msiProt['MSI'] == \"MSI-H\"\n",
    "other = msiProt['MSI'] != \"MSI-H\"\n",
    "highMSI = msiProt[high]\n",
    "otherMSI = msiProt[other]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Find which genes are upregulated in each partition</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ACLY', 'ACY3', 'ADD1', 'ADD3', 'ADO', 'AGO2', 'AGR2', 'ALDH18A1', 'ALKBH4', 'AP1G2', 'AP1M2', 'ARFIP1', 'ARFIP2', 'ARHGAP23', 'ARHGEF40', 'ARMCX1', 'ARMCX2', 'ARPC1A', 'BAG3', 'CACYBP', 'CAND2', 'CBFB', 'CCNA2', 'CCT3', 'CCT5', 'CD109', 'CD2AP', 'CDC42BPB', 'CDC42EP4', 'CELF2', 'CEP112', 'CFI', 'CHAF1A', 'CHAF1B', 'CIC', 'CLPP', 'CLPX', 'CRABP2', 'DCK', 'DDAH1', 'DDAH2', 'DHFR', 'DIP2C', 'DPF2', 'DST', 'DTYMK', 'EFL1', 'EHBP1', 'EIF4E3', 'EML1', 'EPM2AIP1', 'EXOC6B', 'FABP3', 'FANCI', 'FEN1', 'FGD5', 'FNBP1L', 'FOXJ2', 'GALNT7', 'GCC2', 'GCHFR', 'GNPNAT1', 'GOLM1', 'GOLPH3L', 'GPRASP2', 'GSN', 'GSTM2', 'HDAC5', 'HDGFL3', 'HECA', 'HMBS', 'HMGCS1', 'HSPA12A', 'HSPA12B', 'HSPB8', 'IDI1', 'IQSEC1', 'IREB2', 'IRF2BPL', 'IRS1', 'JAK1', 'KIF13A', 'KIF4A', 'KIF7', 'L3MBTL3', 'LAD1', 'LAMB2', 'LDAH', 'LDHA', 'LIG1', 'LLGL1', 'LRBA', 'LRP1', 'LTN1', 'MAOB', 'MAP1B', 'MAP1S', 'MAP4', 'MAP7D1', 'MATN2', 'MBD2', 'MLH1', 'MMP15', 'MMRN2', 'MRE11', 'MVD', 'NCAPG2', 'NCOA1', 'NEK6', 'NES', 'NHSL2', 'NMNAT3', 'NMT2', 'NOTCH2', 'NR2F1', 'NR2F2', 'NT5DC1', 'OCLN', 'OLFML3', 'PALD1', 'PARVG', 'PATJ', 'PBK', 'PBX1', 'PDSS2', 'PFKM', 'PGP', 'PJA1', 'PLCD1', 'PLEC', 'PLEKHA6', 'PMS1', 'PMS2', 'PNMA1', 'PPP1R14A', 'PPP2R3A', 'PRKAR1A', 'PRNP', 'PRTFDC1', 'PSMA1', 'PSMA4', 'PTPRK', 'RAB12', 'RAD50', 'RAP1GAP', 'RASSF2', 'RBM47', 'RBP1', 'RNLS', 'ROCK1', 'ROCK2', 'RPL22L1', 'RPS19', 'RPS6KA1', 'RRM1', 'RSPO1', 'SALL2', 'SAR1B', 'SCP2', 'SCRN1', 'SEPT10', 'SEPT3', 'SEPT5', 'SETD1B', 'SH3PXD2A', 'SKIV2L', 'SMC2', 'SMC4', 'SNRK', 'SNTB2', 'SNX1', 'SOWAHC', 'SPDEF', 'SPTBN1', 'SSNA1', 'STK38', 'STK38L', 'STXBP1', 'SYAP1', 'TACC1', 'TANC2', 'TBX3', 'TCEAL5', 'TECPR2', 'TGFBR3', 'THADA', 'THBS3', 'THUMPD3', 'TIMP2', 'TJAP1', 'TJP2', 'TJP3', 'TMF1', 'TOM1L1', 'TRAP1', 'TRIM2', 'TRIO', 'TRIP6', 'TSC22D2', 'TTC37', 'TUBB2B', 'UAP1', 'UBE2E2', 'UGDH', 'UHRF1', 'VWA5A', 'WDR55', 'YWHAQ', 'ZNF423', 'ZNF503', 'ZSCAN18']\n"
     ]
    }
   ],
   "source": [
    "genes_passing = []\n",
    "genes = highMSI.columns[1:]\n",
    "threshold = .05 / len(genes)\n",
    "for gene in genes:\n",
    "    high_abundance = highMSI[gene]\n",
    "    other_abundance = otherMSI[gene]\n",
    "    pvalue = stats.ttest_ind(high_abundance, other_abundance, nan_policy='omit').pvalue\n",
    "    if pvalue < threshold:\n",
    "        genes_passing.append(gene)\n",
    "print(genes_passing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "('[-1.01 -0.51 -0.93 -0.03 -1.52 -0.05 -0.53  0.28 -0.4   1.55 -0.45  0.41\\n -0.42 -0.39 -0.67  0.19  0.48  1.44 -0.05 -0.79 -0.94 -0.53 -0.51 -0.47\\n  0.14 -0.06 -0.89] not in index', 'occurred at index A1BG')",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-12-f6f0af9b0dfa>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m#TODO: how to pass otherMSI column into apply function? otherMSI[x] doesn't work\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mpassing2\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mhighMSI\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mhighMSI\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mstats\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mttest_ind\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0motherMSI\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m#this is compaing the same column against itself, so it's not doing anything right now\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[0mpassing2\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36mapply\u001b[1;34m(self, func, axis, broadcast, raw, reduce, result_type, args, **kwds)\u001b[0m\n\u001b[0;32m   6002\u001b[0m                          \u001b[0margs\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   6003\u001b[0m                          kwds=kwds)\n\u001b[1;32m-> 6004\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   6005\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   6006\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mapplymap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mget_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    316\u001b[0m                                       *self.args, **self.kwds)\n\u001b[0;32m    317\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 318\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0msuper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mFrameRowApply\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    319\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    320\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mapply_broadcast\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mget_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    140\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_raw\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    141\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 142\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_standard\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    143\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    144\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mapply_empty_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mapply_standard\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    246\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    247\u001b[0m         \u001b[1;31m# compute the result using the series generator\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 248\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply_series_generator\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    249\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    250\u001b[0m         \u001b[1;31m# wrap results\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\apply.py\u001b[0m in \u001b[0;36mapply_series_generator\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    275\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    276\u001b[0m                 \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mv\u001b[0m \u001b[1;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mseries_gen\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 277\u001b[1;33m                     \u001b[0mresults\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mv\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    278\u001b[0m                     \u001b[0mkeys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mv\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    279\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-12-f6f0af9b0dfa>\u001b[0m in \u001b[0;36m<lambda>\u001b[1;34m(x)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;31m#TODO: how to pass otherMSI column into apply function? otherMSI[x] doesn't work\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mpassing2\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mhighMSI\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mhighMSI\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mstats\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mttest_ind\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0motherMSI\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m#this is compaing the same column against itself, so it's not doing anything right now\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[0mpassing2\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   2677\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mSeries\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mIndex\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2678\u001b[0m             \u001b[1;31m# either boolean or fancy integer index\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2679\u001b[1;33m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_array\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2680\u001b[0m         \u001b[1;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mDataFrame\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2681\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_getitem_frame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\frame.py\u001b[0m in \u001b[0;36m_getitem_array\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   2721\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_take\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2722\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 2723\u001b[1;33m             \u001b[0mindexer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_convert_to_indexer\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   2724\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_take\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   2725\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\pandas\\core\\indexing.py\u001b[0m in \u001b[0;36m_convert_to_indexer\u001b[1;34m(self, obj, axis, is_setter)\u001b[0m\n\u001b[0;32m   1325\u001b[0m                 \u001b[1;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0many\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1326\u001b[0m                     raise KeyError('{mask} not in index'\n\u001b[1;32m-> 1327\u001b[1;33m                                    .format(mask=objarr[mask]))\n\u001b[0m\u001b[0;32m   1328\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1329\u001b[0m                 \u001b[1;32mreturn\u001b[0m \u001b[0mcom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_values_from_object\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: ('[-1.01 -0.51 -0.93 -0.03 -1.52 -0.05 -0.53  0.28 -0.4   1.55 -0.45  0.41\\n -0.42 -0.39 -0.67  0.19  0.48  1.44 -0.05 -0.79 -0.94 -0.53 -0.51 -0.47\\n  0.14 -0.06 -0.89] not in index', 'occurred at index A1BG')"
     ]
    }
   ],
   "source": [
    "#TODO: how to pass otherMSI column into apply function? otherMSI[x] doesn't work. Should I combine the dataframes?\n",
    "passing2 = highMSI[highMSI.columns[1:]].apply(lambda x:stats.ttest_ind(x, otherMSI[x])) \n",
    "passing2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<b>Then use the genes that are up-regulated in these partitions to do a gene set enrichment analysis</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Z-score</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Terpenoid backbone biosynthesis_Homo sapiens_h...</td>\n",
       "      <td>4/22</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.008267</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>0.024984</td>\n",
       "      <td>-1.811406</td>\n",
       "      <td>17.738499</td>\n",
       "      <td>IDI1;HMGCS1;PDSS2;MVD</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pathogenic Escherichia coli infection_Homo sap...</td>\n",
       "      <td>5/55</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.014509</td>\n",
       "      <td>0.000483</td>\n",
       "      <td>0.035724</td>\n",
       "      <td>-1.823937</td>\n",
       "      <td>15.571069</td>\n",
       "      <td>OCLN;ROCK1;YWHAQ;ROCK2;ARPC1A</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mismatch repair_Homo sapiens_hsa03430</td>\n",
       "      <td>3/23</td>\n",
       "      <td>0.001400</td>\n",
       "      <td>0.069059</td>\n",
       "      <td>0.002778</td>\n",
       "      <td>0.137051</td>\n",
       "      <td>-1.499972</td>\n",
       "      <td>9.856899</td>\n",
       "      <td>LIG1;PMS2;MLH1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Metabolic pathways_Homo sapiens_hsa01100</td>\n",
       "      <td>21/1239</td>\n",
       "      <td>0.009633</td>\n",
       "      <td>0.248834</td>\n",
       "      <td>0.028499</td>\n",
       "      <td>0.527227</td>\n",
       "      <td>-1.912256</td>\n",
       "      <td>8.877714</td>\n",
       "      <td>GALNT7;IDI1;RRM1;DTYMK;HMGCS1;MAOB;UAP1;NMNAT3...</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Amino sugar and nucleotide sugar metabolism_Ho...</td>\n",
       "      <td>3/48</td>\n",
       "      <td>0.011442</td>\n",
       "      <td>0.248834</td>\n",
       "      <td>0.018288</td>\n",
       "      <td>0.495007</td>\n",
       "      <td>-1.621962</td>\n",
       "      <td>7.250948</td>\n",
       "      <td>UGDH;GNPNAT1;UAP1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tight junction_Homo sapiens_hsa04530</td>\n",
       "      <td>5/139</td>\n",
       "      <td>0.011769</td>\n",
       "      <td>0.248834</td>\n",
       "      <td>0.020830</td>\n",
       "      <td>0.495007</td>\n",
       "      <td>-1.554297</td>\n",
       "      <td>6.904605</td>\n",
       "      <td>OCLN;TJAP1;LLGL1;TJP3;TJP2</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Fanconi anemia pathway_Homo sapiens_hsa03460</td>\n",
       "      <td>3/53</td>\n",
       "      <td>0.014959</td>\n",
       "      <td>0.276738</td>\n",
       "      <td>0.023412</td>\n",
       "      <td>0.495007</td>\n",
       "      <td>-1.485949</td>\n",
       "      <td>6.244631</td>\n",
       "      <td>FANCI;PMS2;MLH1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Non-homologous end-joining_Homo sapiens_hsa03450</td>\n",
       "      <td>2/13</td>\n",
       "      <td>0.006873</td>\n",
       "      <td>0.248834</td>\n",
       "      <td>0.011441</td>\n",
       "      <td>0.423323</td>\n",
       "      <td>-1.101071</td>\n",
       "      <td>5.483427</td>\n",
       "      <td>FEN1;RAD50</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Shigellosis_Homo sapiens_hsa05131</td>\n",
       "      <td>3/65</td>\n",
       "      <td>0.025633</td>\n",
       "      <td>0.421519</td>\n",
       "      <td>0.038566</td>\n",
       "      <td>0.634195</td>\n",
       "      <td>-1.454109</td>\n",
       "      <td>5.327678</td>\n",
       "      <td>ROCK1;ROCK2;ARPC1A</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AMPK signaling pathway_Homo sapiens_hsa04152</td>\n",
       "      <td>4/124</td>\n",
       "      <td>0.033300</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.052046</td>\n",
       "      <td>0.676213</td>\n",
       "      <td>-1.561928</td>\n",
       "      <td>5.314006</td>\n",
       "      <td>CCNA2;IRS1;PPP2R3A;PFKM</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RNA degradation_Homo sapiens_hsa03018</td>\n",
       "      <td>3/77</td>\n",
       "      <td>0.039470</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.057638</td>\n",
       "      <td>0.676213</td>\n",
       "      <td>-1.549493</td>\n",
       "      <td>5.008277</td>\n",
       "      <td>TTC37;SKIV2L;PFKM</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Viral carcinogenesis_Homo sapiens_hsa05203</td>\n",
       "      <td>5/205</td>\n",
       "      <td>0.050733</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.080922</td>\n",
       "      <td>0.704498</td>\n",
       "      <td>-1.600770</td>\n",
       "      <td>4.772188</td>\n",
       "      <td>CCNA2;HDAC5;GSN;YWHAQ;JAK1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Focal adhesion_Homo sapiens_hsa04510</td>\n",
       "      <td>5/202</td>\n",
       "      <td>0.048177</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.077147</td>\n",
       "      <td>0.704498</td>\n",
       "      <td>-1.569970</td>\n",
       "      <td>4.761512</td>\n",
       "      <td>ROCK1;PARVG;ROCK2;LAMB2;THBS3</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Bacterial invasion of epithelial cells_Homo sa...</td>\n",
       "      <td>3/78</td>\n",
       "      <td>0.040764</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.059397</td>\n",
       "      <td>0.676213</td>\n",
       "      <td>-1.448132</td>\n",
       "      <td>4.633956</td>\n",
       "      <td>ARPC1A;SEPT3;CD2AP</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Salmonella infection_Homo sapiens_hsa05132</td>\n",
       "      <td>3/86</td>\n",
       "      <td>0.051869</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.074358</td>\n",
       "      <td>0.704498</td>\n",
       "      <td>-1.248516</td>\n",
       "      <td>3.694394</td>\n",
       "      <td>ROCK1;ROCK2;ARPC1A</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DNA replication_Homo sapiens_hsa03030</td>\n",
       "      <td>2/36</td>\n",
       "      <td>0.047981</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.065158</td>\n",
       "      <td>0.688809</td>\n",
       "      <td>-1.144719</td>\n",
       "      <td>3.476464</td>\n",
       "      <td>FEN1;LIG1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Pyrimidine metabolism_Homo sapiens_hsa00240</td>\n",
       "      <td>3/105</td>\n",
       "      <td>0.083330</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.115687</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.270782</td>\n",
       "      <td>3.157827</td>\n",
       "      <td>DTYMK;RRM1;DCK</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Base excision repair_Homo sapiens_hsa03410</td>\n",
       "      <td>2/33</td>\n",
       "      <td>0.040978</td>\n",
       "      <td>0.451568</td>\n",
       "      <td>0.056314</td>\n",
       "      <td>0.676213</td>\n",
       "      <td>-0.921840</td>\n",
       "      <td>2.945012</td>\n",
       "      <td>FEN1;LIG1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PI3K-Akt signaling pathway_Homo sapiens_hsa04151</td>\n",
       "      <td>6/341</td>\n",
       "      <td>0.117349</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.181685</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.315165</td>\n",
       "      <td>2.817878</td>\n",
       "      <td>YWHAQ;IRS1;LAMB2;PPP2R3A;JAK1;THBS3</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Malaria_Homo sapiens_hsa05144</td>\n",
       "      <td>2/49</td>\n",
       "      <td>0.082576</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.108000</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.124668</td>\n",
       "      <td>2.804969</td>\n",
       "      <td>LRP1;THBS3</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Thyroid hormone signaling pathway_Homo sapiens...</td>\n",
       "      <td>3/118</td>\n",
       "      <td>0.108551</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.147958</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.190757</td>\n",
       "      <td>2.644121</td>\n",
       "      <td>NOTCH2;NCOA1;PLCD1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Vascular smooth muscle contraction_Homo sapien...</td>\n",
       "      <td>3/120</td>\n",
       "      <td>0.112668</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.153164</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.138126</td>\n",
       "      <td>2.484887</td>\n",
       "      <td>PPP1R14A;ROCK1;ROCK2</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Leukocyte transendothelial migration_Homo sapi...</td>\n",
       "      <td>3/118</td>\n",
       "      <td>0.108551</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.147958</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.110863</td>\n",
       "      <td>2.466713</td>\n",
       "      <td>OCLN;ROCK1;ROCK2</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Sphingolipid signaling pathway_Homo sapiens_hs...</td>\n",
       "      <td>3/120</td>\n",
       "      <td>0.112668</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.153164</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.107819</td>\n",
       "      <td>2.418718</td>\n",
       "      <td>ROCK1;ROCK2;PPP2R3A</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Arginine and proline metabolism_Homo sapiens_h...</td>\n",
       "      <td>2/50</td>\n",
       "      <td>0.085484</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.111549</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-0.980809</td>\n",
       "      <td>2.412233</td>\n",
       "      <td>MAOB;ALDH18A1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Epstein-Barr virus infection_Homo sapiens_hsa0...</td>\n",
       "      <td>4/202</td>\n",
       "      <td>0.135573</td>\n",
       "      <td>0.595909</td>\n",
       "      <td>0.190897</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-1.182318</td>\n",
       "      <td>2.362556</td>\n",
       "      <td>CCNA2;HDAC5;YWHAQ;JAK1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Proteasome_Homo sapiens_hsa03050</td>\n",
       "      <td>2/44</td>\n",
       "      <td>0.068527</td>\n",
       "      <td>0.563440</td>\n",
       "      <td>0.090748</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-0.859322</td>\n",
       "      <td>2.303443</td>\n",
       "      <td>PSMA4;PSMA1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Glutathione metabolism_Homo sapiens_hsa00480</td>\n",
       "      <td>2/52</td>\n",
       "      <td>0.091391</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.118739</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-0.919900</td>\n",
       "      <td>2.200966</td>\n",
       "      <td>GSTM2;RRM1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>mTOR signaling pathway_Homo sapiens_hsa04150</td>\n",
       "      <td>2/60</td>\n",
       "      <td>0.116110</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.148553</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-0.914479</td>\n",
       "      <td>1.969070</td>\n",
       "      <td>IRS1;RPS6KA1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Regulation of actin cytoskeleton_Homo sapiens_...</td>\n",
       "      <td>4/214</td>\n",
       "      <td>0.157030</td>\n",
       "      <td>0.595909</td>\n",
       "      <td>0.218326</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>-0.981083</td>\n",
       "      <td>1.816296</td>\n",
       "      <td>GSN;ROCK1;ROCK2;ARPC1A</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>Cocaine addiction_Homo sapiens_hsa05030</td>\n",
       "      <td>1/49</td>\n",
       "      <td>0.381634</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.426047</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.006608</td>\n",
       "      <td>-1.932953</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Type II diabetes mellitus_Homo sapiens_hsa04930</td>\n",
       "      <td>1/48</td>\n",
       "      <td>0.375530</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.419616</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.983015</td>\n",
       "      <td>-1.942196</td>\n",
       "      <td>IRS1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Prion diseases_Homo sapiens_hsa05020</td>\n",
       "      <td>1/35</td>\n",
       "      <td>0.290516</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.329245</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.618875</td>\n",
       "      <td>-2.001085</td>\n",
       "      <td>PRNP</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Nucleotide excision repair_Homo sapiens_hsa03420</td>\n",
       "      <td>1/47</td>\n",
       "      <td>0.369367</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.413113</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.022735</td>\n",
       "      <td>-2.014571</td>\n",
       "      <td>LIG1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Valine, leucine and isoleucine degradation_Hom...</td>\n",
       "      <td>1/48</td>\n",
       "      <td>0.375530</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.419616</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.111519</td>\n",
       "      <td>-2.068055</td>\n",
       "      <td>HMGCS1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>Pentose phosphate pathway_Homo sapiens_hsa00030</td>\n",
       "      <td>1/29</td>\n",
       "      <td>0.247487</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.282979</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.529748</td>\n",
       "      <td>-2.136134</td>\n",
       "      <td>PFKM</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Galactose metabolism_Homo sapiens_hsa00052</td>\n",
       "      <td>1/30</td>\n",
       "      <td>0.254835</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.290903</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.598651</td>\n",
       "      <td>-2.185578</td>\n",
       "      <td>PFKM</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>Tryptophan metabolism_Homo sapiens_hsa00380</td>\n",
       "      <td>1/40</td>\n",
       "      <td>0.324497</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.365538</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.953146</td>\n",
       "      <td>-2.198222</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Cysteine and methionine metabolism_Homo sapien...</td>\n",
       "      <td>1/45</td>\n",
       "      <td>0.356859</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.399892</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.160011</td>\n",
       "      <td>-2.225708</td>\n",
       "      <td>LDHA</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Glycine, serine and threonine metabolism_Homo ...</td>\n",
       "      <td>1/40</td>\n",
       "      <td>0.324497</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.365538</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.991779</td>\n",
       "      <td>-2.241703</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Porphyrin and chlorophyll metabolism_Homo sapi...</td>\n",
       "      <td>1/42</td>\n",
       "      <td>0.337632</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.379507</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.092129</td>\n",
       "      <td>-2.271631</td>\n",
       "      <td>HMBS</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>Pyruvate metabolism_Homo sapiens_hsa00620</td>\n",
       "      <td>1/40</td>\n",
       "      <td>0.324497</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.365538</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.093978</td>\n",
       "      <td>-2.356726</td>\n",
       "      <td>LDHA</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>Tyrosine metabolism_Homo sapiens_hsa00350</td>\n",
       "      <td>1/35</td>\n",
       "      <td>0.290516</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.329245</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.999179</td>\n",
       "      <td>-2.471177</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism_Homo s...</td>\n",
       "      <td>1/28</td>\n",
       "      <td>0.240068</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.274968</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.737792</td>\n",
       "      <td>-2.479542</td>\n",
       "      <td>PGP</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>Butanoate metabolism_Homo sapiens_hsa00650</td>\n",
       "      <td>1/28</td>\n",
       "      <td>0.240068</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.274968</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.742925</td>\n",
       "      <td>-2.486866</td>\n",
       "      <td>HMGCS1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Nicotinate and nicotinamide metabolism_Homo sa...</td>\n",
       "      <td>1/29</td>\n",
       "      <td>0.247487</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.282979</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.844405</td>\n",
       "      <td>-2.575519</td>\n",
       "      <td>NMNAT3</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Homologous recombination_Homo sapiens_hsa03440</td>\n",
       "      <td>1/29</td>\n",
       "      <td>0.247487</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.282979</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.853628</td>\n",
       "      <td>-2.588398</td>\n",
       "      <td>RAD50</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Fructose and mannose metabolism_Homo sapiens_h...</td>\n",
       "      <td>1/32</td>\n",
       "      <td>0.269317</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.306493</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.977494</td>\n",
       "      <td>-2.594209</td>\n",
       "      <td>PFKM</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Propanoate metabolism_Homo sapiens_hsa00640</td>\n",
       "      <td>1/32</td>\n",
       "      <td>0.269317</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.306493</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.984557</td>\n",
       "      <td>-2.603474</td>\n",
       "      <td>LDHA</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Pentose and glucuronate interconversions_Homo ...</td>\n",
       "      <td>1/36</td>\n",
       "      <td>0.297446</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.336664</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.222187</td>\n",
       "      <td>-2.694453</td>\n",
       "      <td>UGDH</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Citrate cycle (TCA cycle)_Homo sapiens_hsa00020</td>\n",
       "      <td>1/30</td>\n",
       "      <td>0.254835</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.290903</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.043919</td>\n",
       "      <td>-2.794322</td>\n",
       "      <td>ACLY</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>Mucin type O-Glycan biosynthesis_Homo sapiens_...</td>\n",
       "      <td>1/31</td>\n",
       "      <td>0.262111</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.298741</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.089837</td>\n",
       "      <td>-2.798263</td>\n",
       "      <td>GALNT7</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Ascorbate and aldarate metabolism_Homo sapiens...</td>\n",
       "      <td>1/27</td>\n",
       "      <td>0.232575</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.266868</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.981127</td>\n",
       "      <td>-2.889557</td>\n",
       "      <td>UGDH</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Histidine metabolism_Homo sapiens_hsa00340</td>\n",
       "      <td>1/24</td>\n",
       "      <td>0.209654</td>\n",
       "      <td>0.656928</td>\n",
       "      <td>0.242029</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>1.935621</td>\n",
       "      <td>-3.024017</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>Phenylalanine metabolism_Homo sapiens_hsa00360</td>\n",
       "      <td>1/17</td>\n",
       "      <td>0.153488</td>\n",
       "      <td>0.595909</td>\n",
       "      <td>0.180794</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.063751</td>\n",
       "      <td>-3.867749</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>One carbon pool by folate_Homo sapiens_hsa00670</td>\n",
       "      <td>1/20</td>\n",
       "      <td>0.178029</td>\n",
       "      <td>0.627341</td>\n",
       "      <td>0.207614</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.334290</td>\n",
       "      <td>-4.028536</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Primary bile acid biosynthesis_Homo sapiens_hs...</td>\n",
       "      <td>1/17</td>\n",
       "      <td>0.153488</td>\n",
       "      <td>0.595909</td>\n",
       "      <td>0.180794</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.367420</td>\n",
       "      <td>-4.436867</td>\n",
       "      <td>SCP2</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>Folate biosynthesis_Homo sapiens_hsa00790</td>\n",
       "      <td>1/14</td>\n",
       "      <td>0.128217</td>\n",
       "      <td>0.595909</td>\n",
       "      <td>0.153079</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.169316</td>\n",
       "      <td>-4.455844</td>\n",
       "      <td>DHFR</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Taurine and hypotaurine metabolism_Homo sapien...</td>\n",
       "      <td>1/11</td>\n",
       "      <td>0.102196</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.124438</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.150657</td>\n",
       "      <td>-4.905357</td>\n",
       "      <td>ADO</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Synthesis and degradation of ketone bodies_Hom...</td>\n",
       "      <td>1/10</td>\n",
       "      <td>0.093352</td>\n",
       "      <td>0.578921</td>\n",
       "      <td>0.114681</td>\n",
       "      <td>0.732334</td>\n",
       "      <td>2.102183</td>\n",
       "      <td>-4.985079</td>\n",
       "      <td>HMGCS1</td>\n",
       "      <td>KEGG_2016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>148 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  Term  Overlap   P-value  \\\n",
       "0    Terpenoid backbone biosynthesis_Homo sapiens_h...     4/22  0.000056   \n",
       "1    Pathogenic Escherichia coli infection_Homo sap...     5/55  0.000196   \n",
       "2                Mismatch repair_Homo sapiens_hsa03430     3/23  0.001400   \n",
       "3             Metabolic pathways_Homo sapiens_hsa01100  21/1239  0.009633   \n",
       "4    Amino sugar and nucleotide sugar metabolism_Ho...     3/48  0.011442   \n",
       "5                 Tight junction_Homo sapiens_hsa04530    5/139  0.011769   \n",
       "6         Fanconi anemia pathway_Homo sapiens_hsa03460     3/53  0.014959   \n",
       "7     Non-homologous end-joining_Homo sapiens_hsa03450     2/13  0.006873   \n",
       "8                    Shigellosis_Homo sapiens_hsa05131     3/65  0.025633   \n",
       "9         AMPK signaling pathway_Homo sapiens_hsa04152    4/124  0.033300   \n",
       "10               RNA degradation_Homo sapiens_hsa03018     3/77  0.039470   \n",
       "11          Viral carcinogenesis_Homo sapiens_hsa05203    5/205  0.050733   \n",
       "12                Focal adhesion_Homo sapiens_hsa04510    5/202  0.048177   \n",
       "13   Bacterial invasion of epithelial cells_Homo sa...     3/78  0.040764   \n",
       "14          Salmonella infection_Homo sapiens_hsa05132     3/86  0.051869   \n",
       "15               DNA replication_Homo sapiens_hsa03030     2/36  0.047981   \n",
       "16         Pyrimidine metabolism_Homo sapiens_hsa00240    3/105  0.083330   \n",
       "17          Base excision repair_Homo sapiens_hsa03410     2/33  0.040978   \n",
       "18    PI3K-Akt signaling pathway_Homo sapiens_hsa04151    6/341  0.117349   \n",
       "19                       Malaria_Homo sapiens_hsa05144     2/49  0.082576   \n",
       "20   Thyroid hormone signaling pathway_Homo sapiens...    3/118  0.108551   \n",
       "21   Vascular smooth muscle contraction_Homo sapien...    3/120  0.112668   \n",
       "22   Leukocyte transendothelial migration_Homo sapi...    3/118  0.108551   \n",
       "23   Sphingolipid signaling pathway_Homo sapiens_hs...    3/120  0.112668   \n",
       "24   Arginine and proline metabolism_Homo sapiens_h...     2/50  0.085484   \n",
       "25   Epstein-Barr virus infection_Homo sapiens_hsa0...    4/202  0.135573   \n",
       "26                    Proteasome_Homo sapiens_hsa03050     2/44  0.068527   \n",
       "27        Glutathione metabolism_Homo sapiens_hsa00480     2/52  0.091391   \n",
       "28        mTOR signaling pathway_Homo sapiens_hsa04150     2/60  0.116110   \n",
       "29   Regulation of actin cytoskeleton_Homo sapiens_...    4/214  0.157030   \n",
       "..                                                 ...      ...       ...   \n",
       "118            Cocaine addiction_Homo sapiens_hsa05030     1/49  0.381634   \n",
       "119    Type II diabetes mellitus_Homo sapiens_hsa04930     1/48  0.375530   \n",
       "120               Prion diseases_Homo sapiens_hsa05020     1/35  0.290516   \n",
       "121   Nucleotide excision repair_Homo sapiens_hsa03420     1/47  0.369367   \n",
       "122  Valine, leucine and isoleucine degradation_Hom...     1/48  0.375530   \n",
       "123    Pentose phosphate pathway_Homo sapiens_hsa00030     1/29  0.247487   \n",
       "124         Galactose metabolism_Homo sapiens_hsa00052     1/30  0.254835   \n",
       "125        Tryptophan metabolism_Homo sapiens_hsa00380     1/40  0.324497   \n",
       "126  Cysteine and methionine metabolism_Homo sapien...     1/45  0.356859   \n",
       "127  Glycine, serine and threonine metabolism_Homo ...     1/40  0.324497   \n",
       "128  Porphyrin and chlorophyll metabolism_Homo sapi...     1/42  0.337632   \n",
       "129          Pyruvate metabolism_Homo sapiens_hsa00620     1/40  0.324497   \n",
       "130          Tyrosine metabolism_Homo sapiens_hsa00350     1/35  0.290516   \n",
       "131  Glyoxylate and dicarboxylate metabolism_Homo s...     1/28  0.240068   \n",
       "132         Butanoate metabolism_Homo sapiens_hsa00650     1/28  0.240068   \n",
       "133  Nicotinate and nicotinamide metabolism_Homo sa...     1/29  0.247487   \n",
       "134     Homologous recombination_Homo sapiens_hsa03440     1/29  0.247487   \n",
       "135  Fructose and mannose metabolism_Homo sapiens_h...     1/32  0.269317   \n",
       "136        Propanoate metabolism_Homo sapiens_hsa00640     1/32  0.269317   \n",
       "137  Pentose and glucuronate interconversions_Homo ...     1/36  0.297446   \n",
       "138    Citrate cycle (TCA cycle)_Homo sapiens_hsa00020     1/30  0.254835   \n",
       "139  Mucin type O-Glycan biosynthesis_Homo sapiens_...     1/31  0.262111   \n",
       "140  Ascorbate and aldarate metabolism_Homo sapiens...     1/27  0.232575   \n",
       "141         Histidine metabolism_Homo sapiens_hsa00340     1/24  0.209654   \n",
       "142     Phenylalanine metabolism_Homo sapiens_hsa00360     1/17  0.153488   \n",
       "143    One carbon pool by folate_Homo sapiens_hsa00670     1/20  0.178029   \n",
       "144  Primary bile acid biosynthesis_Homo sapiens_hs...     1/17  0.153488   \n",
       "145          Folate biosynthesis_Homo sapiens_hsa00790     1/14  0.128217   \n",
       "146  Taurine and hypotaurine metabolism_Homo sapien...     1/11  0.102196   \n",
       "147  Synthesis and degradation of ketone bodies_Hom...     1/10  0.093352   \n",
       "\n",
       "     Adjusted P-value  Old P-value  Old Adjusted P-value   Z-score  \\\n",
       "0            0.008267     0.000169              0.024984 -1.811406   \n",
       "1            0.014509     0.000483              0.035724 -1.823937   \n",
       "2            0.069059     0.002778              0.137051 -1.499972   \n",
       "3            0.248834     0.028499              0.527227 -1.912256   \n",
       "4            0.248834     0.018288              0.495007 -1.621962   \n",
       "5            0.248834     0.020830              0.495007 -1.554297   \n",
       "6            0.276738     0.023412              0.495007 -1.485949   \n",
       "7            0.248834     0.011441              0.423323 -1.101071   \n",
       "8            0.421519     0.038566              0.634195 -1.454109   \n",
       "9            0.451568     0.052046              0.676213 -1.561928   \n",
       "10           0.451568     0.057638              0.676213 -1.549493   \n",
       "11           0.451568     0.080922              0.704498 -1.600770   \n",
       "12           0.451568     0.077147              0.704498 -1.569970   \n",
       "13           0.451568     0.059397              0.676213 -1.448132   \n",
       "14           0.451568     0.074358              0.704498 -1.248516   \n",
       "15           0.451568     0.065158              0.688809 -1.144719   \n",
       "16           0.578921     0.115687              0.732334 -1.270782   \n",
       "17           0.451568     0.056314              0.676213 -0.921840   \n",
       "18           0.578921     0.181685              0.732334 -1.315165   \n",
       "19           0.578921     0.108000              0.732334 -1.124668   \n",
       "20           0.578921     0.147958              0.732334 -1.190757   \n",
       "21           0.578921     0.153164              0.732334 -1.138126   \n",
       "22           0.578921     0.147958              0.732334 -1.110863   \n",
       "23           0.578921     0.153164              0.732334 -1.107819   \n",
       "24           0.578921     0.111549              0.732334 -0.980809   \n",
       "25           0.595909     0.190897              0.732334 -1.182318   \n",
       "26           0.563440     0.090748              0.732334 -0.859322   \n",
       "27           0.578921     0.118739              0.732334 -0.919900   \n",
       "28           0.578921     0.148553              0.732334 -0.914479   \n",
       "29           0.595909     0.218326              0.732334 -0.981083   \n",
       "..                ...          ...                   ...       ...   \n",
       "118          0.656928     0.426047              0.732334  2.006608   \n",
       "119          0.656928     0.419616              0.732334  1.983015   \n",
       "120          0.656928     0.329245              0.732334  1.618875   \n",
       "121          0.656928     0.413113              0.732334  2.022735   \n",
       "122          0.656928     0.419616              0.732334  2.111519   \n",
       "123          0.656928     0.282979              0.732334  1.529748   \n",
       "124          0.656928     0.290903              0.732334  1.598651   \n",
       "125          0.656928     0.365538              0.732334  1.953146   \n",
       "126          0.656928     0.399892              0.732334  2.160011   \n",
       "127          0.656928     0.365538              0.732334  1.991779   \n",
       "128          0.656928     0.379507              0.732334  2.092129   \n",
       "129          0.656928     0.365538              0.732334  2.093978   \n",
       "130          0.656928     0.329245              0.732334  1.999179   \n",
       "131          0.656928     0.274968              0.732334  1.737792   \n",
       "132          0.656928     0.274968              0.732334  1.742925   \n",
       "133          0.656928     0.282979              0.732334  1.844405   \n",
       "134          0.656928     0.282979              0.732334  1.853628   \n",
       "135          0.656928     0.306493              0.732334  1.977494   \n",
       "136          0.656928     0.306493              0.732334  1.984557   \n",
       "137          0.656928     0.336664              0.732334  2.222187   \n",
       "138          0.656928     0.290903              0.732334  2.043919   \n",
       "139          0.656928     0.298741              0.732334  2.089837   \n",
       "140          0.656928     0.266868              0.732334  1.981127   \n",
       "141          0.656928     0.242029              0.732334  1.935621   \n",
       "142          0.595909     0.180794              0.732334  2.063751   \n",
       "143          0.627341     0.207614              0.732334  2.334290   \n",
       "144          0.595909     0.180794              0.732334  2.367420   \n",
       "145          0.595909     0.153079              0.732334  2.169316   \n",
       "146          0.578921     0.124438              0.732334  2.150657   \n",
       "147          0.578921     0.114681              0.732334  2.102183   \n",
       "\n",
       "     Combined Score                                              Genes  \\\n",
       "0         17.738499                              IDI1;HMGCS1;PDSS2;MVD   \n",
       "1         15.571069                      OCLN;ROCK1;YWHAQ;ROCK2;ARPC1A   \n",
       "2          9.856899                                     LIG1;PMS2;MLH1   \n",
       "3          8.877714  GALNT7;IDI1;RRM1;DTYMK;HMGCS1;MAOB;UAP1;NMNAT3...   \n",
       "4          7.250948                                  UGDH;GNPNAT1;UAP1   \n",
       "5          6.904605                         OCLN;TJAP1;LLGL1;TJP3;TJP2   \n",
       "6          6.244631                                    FANCI;PMS2;MLH1   \n",
       "7          5.483427                                         FEN1;RAD50   \n",
       "8          5.327678                                 ROCK1;ROCK2;ARPC1A   \n",
       "9          5.314006                            CCNA2;IRS1;PPP2R3A;PFKM   \n",
       "10         5.008277                                  TTC37;SKIV2L;PFKM   \n",
       "11         4.772188                         CCNA2;HDAC5;GSN;YWHAQ;JAK1   \n",
       "12         4.761512                      ROCK1;PARVG;ROCK2;LAMB2;THBS3   \n",
       "13         4.633956                                 ARPC1A;SEPT3;CD2AP   \n",
       "14         3.694394                                 ROCK1;ROCK2;ARPC1A   \n",
       "15         3.476464                                          FEN1;LIG1   \n",
       "16         3.157827                                     DTYMK;RRM1;DCK   \n",
       "17         2.945012                                          FEN1;LIG1   \n",
       "18         2.817878                YWHAQ;IRS1;LAMB2;PPP2R3A;JAK1;THBS3   \n",
       "19         2.804969                                         LRP1;THBS3   \n",
       "20         2.644121                                 NOTCH2;NCOA1;PLCD1   \n",
       "21         2.484887                               PPP1R14A;ROCK1;ROCK2   \n",
       "22         2.466713                                   OCLN;ROCK1;ROCK2   \n",
       "23         2.418718                                ROCK1;ROCK2;PPP2R3A   \n",
       "24         2.412233                                      MAOB;ALDH18A1   \n",
       "25         2.362556                             CCNA2;HDAC5;YWHAQ;JAK1   \n",
       "26         2.303443                                        PSMA4;PSMA1   \n",
       "27         2.200966                                         GSTM2;RRM1   \n",
       "28         1.969070                                       IRS1;RPS6KA1   \n",
       "29         1.816296                             GSN;ROCK1;ROCK2;ARPC1A   \n",
       "..              ...                                                ...   \n",
       "118       -1.932953                                               MAOB   \n",
       "119       -1.942196                                               IRS1   \n",
       "120       -2.001085                                               PRNP   \n",
       "121       -2.014571                                               LIG1   \n",
       "122       -2.068055                                             HMGCS1   \n",
       "123       -2.136134                                               PFKM   \n",
       "124       -2.185578                                               PFKM   \n",
       "125       -2.198222                                               MAOB   \n",
       "126       -2.225708                                               LDHA   \n",
       "127       -2.241703                                               MAOB   \n",
       "128       -2.271631                                               HMBS   \n",
       "129       -2.356726                                               LDHA   \n",
       "130       -2.471177                                               MAOB   \n",
       "131       -2.479542                                                PGP   \n",
       "132       -2.486866                                             HMGCS1   \n",
       "133       -2.575519                                             NMNAT3   \n",
       "134       -2.588398                                              RAD50   \n",
       "135       -2.594209                                               PFKM   \n",
       "136       -2.603474                                               LDHA   \n",
       "137       -2.694453                                               UGDH   \n",
       "138       -2.794322                                               ACLY   \n",
       "139       -2.798263                                             GALNT7   \n",
       "140       -2.889557                                               UGDH   \n",
       "141       -3.024017                                               MAOB   \n",
       "142       -3.867749                                               MAOB   \n",
       "143       -4.028536                                               DHFR   \n",
       "144       -4.436867                                               SCP2   \n",
       "145       -4.455844                                               DHFR   \n",
       "146       -4.905357                                                ADO   \n",
       "147       -4.985079                                             HMGCS1   \n",
       "\n",
       "      Gene_set  \n",
       "0    KEGG_2016  \n",
       "1    KEGG_2016  \n",
       "2    KEGG_2016  \n",
       "3    KEGG_2016  \n",
       "4    KEGG_2016  \n",
       "5    KEGG_2016  \n",
       "6    KEGG_2016  \n",
       "7    KEGG_2016  \n",
       "8    KEGG_2016  \n",
       "9    KEGG_2016  \n",
       "10   KEGG_2016  \n",
       "11   KEGG_2016  \n",
       "12   KEGG_2016  \n",
       "13   KEGG_2016  \n",
       "14   KEGG_2016  \n",
       "15   KEGG_2016  \n",
       "16   KEGG_2016  \n",
       "17   KEGG_2016  \n",
       "18   KEGG_2016  \n",
       "19   KEGG_2016  \n",
       "20   KEGG_2016  \n",
       "21   KEGG_2016  \n",
       "22   KEGG_2016  \n",
       "23   KEGG_2016  \n",
       "24   KEGG_2016  \n",
       "25   KEGG_2016  \n",
       "26   KEGG_2016  \n",
       "27   KEGG_2016  \n",
       "28   KEGG_2016  \n",
       "29   KEGG_2016  \n",
       "..         ...  \n",
       "118  KEGG_2016  \n",
       "119  KEGG_2016  \n",
       "120  KEGG_2016  \n",
       "121  KEGG_2016  \n",
       "122  KEGG_2016  \n",
       "123  KEGG_2016  \n",
       "124  KEGG_2016  \n",
       "125  KEGG_2016  \n",
       "126  KEGG_2016  \n",
       "127  KEGG_2016  \n",
       "128  KEGG_2016  \n",
       "129  KEGG_2016  \n",
       "130  KEGG_2016  \n",
       "131  KEGG_2016  \n",
       "132  KEGG_2016  \n",
       "133  KEGG_2016  \n",
       "134  KEGG_2016  \n",
       "135  KEGG_2016  \n",
       "136  KEGG_2016  \n",
       "137  KEGG_2016  \n",
       "138  KEGG_2016  \n",
       "139  KEGG_2016  \n",
       "140  KEGG_2016  \n",
       "141  KEGG_2016  \n",
       "142  KEGG_2016  \n",
       "143  KEGG_2016  \n",
       "144  KEGG_2016  \n",
       "145  KEGG_2016  \n",
       "146  KEGG_2016  \n",
       "147  KEGG_2016  \n",
       "\n",
       "[148 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enr = gp.enrichr(gene_list = genes_passing, description='MSI partitions', gene_sets='KEGG_2016', outdir='test/enrichr_kegg',cutoff=.5)\n",
    "enr.res2d"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
