{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 3: Joining dataframes with <code>cptac</code>\n",
    "\n",
    "In this use case, we provide several examples of how to use the built-in <code>cptac</code> functions for joining different dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                    \r"
     ]
    }
   ],
   "source": [
    "import cptac\n",
    "en = cptac.Endometrial()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## General format\n",
    "\n",
    "In all of the join functions, you specify the dataframes you want to join by passing their names to the appropriate parameters in the function call. The function will automatically check that the dataframes whose names you provided are valid for the join function, and print an error message if they aren't.\n",
    "\n",
    "Whenever a column from an -omics dataframe is included in a joined table, the name of the -omics dataframe it came from is joined to the column header, to avoid confusion.\n",
    "\n",
    "If you wish to only include particular columns in the join, pass them to the appropriate parameters in the join function. All such parameters will accept either a single column name as a string, or a list of column name strings. In this use case, we will usually only select specific columns for readability, but you could select the whole dataframe in all these cases, except for the mutations dataframe.\n",
    "\n",
    "The join functions use logic analogous to an SQL INNER JOIN."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_omics_to_omics</code>\n",
    "\n",
    "The <code>join_omics_to_omics</code> function joins two -omics dataframes to each other. Types of -omics data valid for use with this function are acetylproteomics, CNV, phosphoproteomics, phosphoproteomics_gene, proteomics, and transcriptomics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>A2M_proteomics</th>\n",
       "      <th>A2ML1_proteomics</th>\n",
       "      <th>A4GALT_proteomics</th>\n",
       "      <th>AAAS_proteomics</th>\n",
       "      <th>AACS_proteomics</th>\n",
       "      <th>AADAT_proteomics</th>\n",
       "      <th>AAED1_proteomics</th>\n",
       "      <th>AAGAB_proteomics</th>\n",
       "      <th>AAK1_proteomics</th>\n",
       "      <th>...</th>\n",
       "      <th>ZZZ3-S397_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S411_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S420_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S424_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S426_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S468_phosphoproteomics</th>\n",
       "      <th>ZZZ3-S89_phosphoproteomics</th>\n",
       "      <th>ZZZ3-T415_phosphoproteomics</th>\n",
       "      <th>ZZZ3-T418_phosphoproteomics</th>\n",
       "      <th>ZZZ3-Y399_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>-1.180</td>\n",
       "      <td>-0.8630</td>\n",
       "      <td>-0.802</td>\n",
       "      <td>0.222</td>\n",
       "      <td>0.2560</td>\n",
       "      <td>0.6650</td>\n",
       "      <td>1.2800</td>\n",
       "      <td>-0.3390</td>\n",
       "      <td>0.412</td>\n",
       "      <td>-0.664</td>\n",
       "      <td>...</td>\n",
       "      <td>0.18400</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.20500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>-0.685</td>\n",
       "      <td>-1.0700</td>\n",
       "      <td>-0.684</td>\n",
       "      <td>0.984</td>\n",
       "      <td>0.1350</td>\n",
       "      <td>0.3340</td>\n",
       "      <td>1.3000</td>\n",
       "      <td>0.1390</td>\n",
       "      <td>1.330</td>\n",
       "      <td>-0.367</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.17100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.393</td>\n",
       "      <td>-0.17100</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1605</td>\n",
       "      <td>-0.0635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.528</td>\n",
       "      <td>-1.3200</td>\n",
       "      <td>0.435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.2400</td>\n",
       "      <td>1.0400</td>\n",
       "      <td>-0.0213</td>\n",
       "      <td>-0.0479</td>\n",
       "      <td>0.419</td>\n",
       "      <td>-0.500</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-1.670</td>\n",
       "      <td>-1.1900</td>\n",
       "      <td>-0.443</td>\n",
       "      <td>0.243</td>\n",
       "      <td>-0.0993</td>\n",
       "      <td>0.7570</td>\n",
       "      <td>0.7400</td>\n",
       "      <td>-0.9290</td>\n",
       "      <td>0.229</td>\n",
       "      <td>-0.223</td>\n",
       "      <td>...</td>\n",
       "      <td>0.13970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.55900</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>-0.374</td>\n",
       "      <td>-0.0206</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>0.311</td>\n",
       "      <td>0.3750</td>\n",
       "      <td>0.0131</td>\n",
       "      <td>-1.1000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.565</td>\n",
       "      <td>-0.101</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.15875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.06175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.2900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 84211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG_proteomics  A2M_proteomics  A2ML1_proteomics  A4GALT_proteomics  AAAS_proteomics  AACS_proteomics  AADAT_proteomics  AAED1_proteomics  AAGAB_proteomics  AAK1_proteomics  ...  \\\n",
       "Sample_ID                                                                                                                                                                                 ...   \n",
       "S001                -1.180         -0.8630            -0.802              0.222           0.2560           0.6650            1.2800           -0.3390             0.412           -0.664  ...   \n",
       "S002                -0.685         -1.0700            -0.684              0.984           0.1350           0.3340            1.3000            0.1390             1.330           -0.367  ...   \n",
       "S003                -0.528         -1.3200             0.435                NaN          -0.2400           1.0400           -0.0213           -0.0479             0.419           -0.500  ...   \n",
       "S005                -1.670         -1.1900            -0.443              0.243          -0.0993           0.7570            0.7400           -0.9290             0.229           -0.223  ...   \n",
       "S006                -0.374         -0.0206            -0.537              0.311           0.3750           0.0131           -1.1000               NaN             0.565           -0.101  ...   \n",
       "\n",
       "           ZZZ3-S397_phosphoproteomics  ZZZ3-S411_phosphoproteomics  ZZZ3-S420_phosphoproteomics  ZZZ3-S424_phosphoproteomics  ZZZ3-S426_phosphoproteomics  ZZZ3-S468_phosphoproteomics  \\\n",
       "Sample_ID                                                                                                                                                                                 \n",
       "S001                           0.18400                          NaN                          NaN                          NaN                     -0.20500                          NaN   \n",
       "S002                          -0.17100                          NaN                          NaN                       -0.393                     -0.17100                          NaN   \n",
       "S003                               NaN                          NaN                          NaN                          NaN                          NaN                          NaN   \n",
       "S005                           0.13970                          NaN                          NaN                          NaN                     -0.55900                          NaN   \n",
       "S006                          -0.15875                          NaN                          NaN                        0.196                      0.06175                          NaN   \n",
       "\n",
       "           ZZZ3-S89_phosphoproteomics  ZZZ3-T415_phosphoproteomics  ZZZ3-T418_phosphoproteomics  ZZZ3-Y399_phosphoproteomics  \n",
       "Sample_ID                                                                                                                     \n",
       "S001                              NaN                          NaN                          NaN                          NaN  \n",
       "S002                             0.29                          NaN                       0.1605                      -0.0635  \n",
       "S003                              NaN                          NaN                          NaN                          NaN  \n",
       "S005                              NaN                          NaN                          NaN                       0.2980  \n",
       "S006                              NaN                          NaN                          NaN                      -0.2900  \n",
       "\n",
       "[5 rows x 84211 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_and_phos = en.join_omics_to_omics(df1_name=\"proteomics\", df2_name=\"phosphoproteomics\")\n",
    "prot_and_phos.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joining only specific columns.\n",
    "(Note that when a gene is selected from the phosphoproteomics dataframe, data for all sites of the gene are selected. The same is done for acetylproteomics data.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG_proteomics</th>\n",
       "      <th>PIK3CA-S312_phosphoproteomics</th>\n",
       "      <th>PIK3CA-T313_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>-1.180</td>\n",
       "      <td>-0.00615</td>\n",
       "      <td>0.0731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>-0.685</td>\n",
       "      <td>-0.02220</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>-0.528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-1.670</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.8460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>-0.374</td>\n",
       "      <td>0.43600</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG_proteomics  PIK3CA-S312_phosphoproteomics  PIK3CA-T313_phosphoproteomics\n",
       "Sample_ID                                                                               \n",
       "S001                -1.180                       -0.00615                         0.0731\n",
       "S002                -0.685                       -0.02220                            NaN\n",
       "S003                -0.528                            NaN                         0.0830\n",
       "S005                -1.670                            NaN                        -0.8460\n",
       "S006                -0.374                        0.43600                            NaN"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prot_and_phos_selected = en.join_omics_to_omics(\n",
    "    df1_name=\"proteomics\", \n",
    "    df2_name=\"phosphoproteomics\", \n",
    "    genes1=\"A1BG\", \n",
    "    genes2=\"PIK3CA\")\n",
    "\n",
    "prot_and_phos_selected.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_metadata_to_omics</code>\n",
    "\n",
    "The <code>join_metadata_to_omics</code> function joins a metadata dataframe (e.g. clinical or derived_molecular) with an -omics dataframe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Proteomics_Tumor_Normal</th>\n",
       "      <th>Country</th>\n",
       "      <th>Histologic_Grade_FIGO</th>\n",
       "      <th>Myometrial_invasion_Specify</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Treatment_naive</th>\n",
       "      <th>Tumor_purity</th>\n",
       "      <th>Path_Stage_Primary_Tumor-pT</th>\n",
       "      <th>Path_Stage_Reg_Lymph_Nodes-pN</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH_transcriptomics</th>\n",
       "      <th>ZWINT_transcriptomics</th>\n",
       "      <th>ZXDA_transcriptomics</th>\n",
       "      <th>ZXDB_transcriptomics</th>\n",
       "      <th>ZXDC_transcriptomics</th>\n",
       "      <th>ZYG11A_transcriptomics</th>\n",
       "      <th>ZYG11B_transcriptomics</th>\n",
       "      <th>ZYX_transcriptomics</th>\n",
       "      <th>ZZEF1_transcriptomics</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>...</td>\n",
       "      <td>11.06</td>\n",
       "      <td>10.73</td>\n",
       "      <td>8.40</td>\n",
       "      <td>9.78</td>\n",
       "      <td>10.88</td>\n",
       "      <td>5.93</td>\n",
       "      <td>11.52</td>\n",
       "      <td>10.23</td>\n",
       "      <td>11.50</td>\n",
       "      <td>11.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.87</td>\n",
       "      <td>11.43</td>\n",
       "      <td>8.39</td>\n",
       "      <td>9.14</td>\n",
       "      <td>10.38</td>\n",
       "      <td>7.25</td>\n",
       "      <td>11.64</td>\n",
       "      <td>10.64</td>\n",
       "      <td>11.26</td>\n",
       "      <td>11.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pN0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.06</td>\n",
       "      <td>10.13</td>\n",
       "      <td>8.35</td>\n",
       "      <td>9.27</td>\n",
       "      <td>10.46</td>\n",
       "      <td>6.85</td>\n",
       "      <td>11.60</td>\n",
       "      <td>10.21</td>\n",
       "      <td>11.51</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.29</td>\n",
       "      <td>10.41</td>\n",
       "      <td>9.10</td>\n",
       "      <td>9.59</td>\n",
       "      <td>10.15</td>\n",
       "      <td>7.89</td>\n",
       "      <td>11.90</td>\n",
       "      <td>10.21</td>\n",
       "      <td>11.34</td>\n",
       "      <td>11.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>C3L-00098</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>United States</td>\n",
       "      <td>NaN</td>\n",
       "      <td>under 50 %</td>\n",
       "      <td>Serous</td>\n",
       "      <td>YES</td>\n",
       "      <td>Normal</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>pNX</td>\n",
       "      <td>...</td>\n",
       "      <td>10.36</td>\n",
       "      <td>11.24</td>\n",
       "      <td>8.60</td>\n",
       "      <td>9.44</td>\n",
       "      <td>11.80</td>\n",
       "      <td>9.32</td>\n",
       "      <td>11.97</td>\n",
       "      <td>9.77</td>\n",
       "      <td>11.37</td>\n",
       "      <td>12.35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28083 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Patient_ID Proteomics_Tumor_Normal        Country Histologic_Grade_FIGO Myometrial_invasion_Specify Histologic_type Treatment_naive Tumor_purity Path_Stage_Primary_Tumor-pT  \\\n",
       "Sample_ID                                                                                                                                                                                \n",
       "S001       C3L-00006                   Tumor  United States          FIGO grade 1                  under 50 %    Endometrioid             YES       Normal              pT1a (FIGO IA)   \n",
       "S002       C3L-00008                   Tumor  United States          FIGO grade 1                  under 50 %    Endometrioid             YES       Normal              pT1a (FIGO IA)   \n",
       "S003       C3L-00032                   Tumor  United States          FIGO grade 2                  under 50 %    Endometrioid             YES       Normal              pT1a (FIGO IA)   \n",
       "S005       C3L-00090                   Tumor  United States          FIGO grade 2                  under 50 %    Endometrioid             YES       Normal              pT1a (FIGO IA)   \n",
       "S006       C3L-00098                   Tumor  United States                   NaN                  under 50 %          Serous             YES       Normal              pT1a (FIGO IA)   \n",
       "\n",
       "          Path_Stage_Reg_Lymph_Nodes-pN  ... ZWILCH_transcriptomics ZWINT_transcriptomics ZXDA_transcriptomics ZXDB_transcriptomics  ZXDC_transcriptomics  ZYG11A_transcriptomics  \\\n",
       "Sample_ID                                ...                                                                                                                                        \n",
       "S001                                pN0  ...                  11.06                 10.73                 8.40                 9.78                 10.88                    5.93   \n",
       "S002                                pNX  ...                  10.87                 11.43                 8.39                 9.14                 10.38                    7.25   \n",
       "S003                                pN0  ...                  10.06                 10.13                 8.35                 9.27                 10.46                    6.85   \n",
       "S005                                pNX  ...                  10.29                 10.41                 9.10                 9.59                 10.15                    7.89   \n",
       "S006                                pNX  ...                  10.36                 11.24                 8.60                 9.44                 11.80                    9.32   \n",
       "\n",
       "           ZYG11B_transcriptomics ZYX_transcriptomics ZZEF1_transcriptomics ZZZ3_transcriptomics  \n",
       "Sample_ID                                                                                         \n",
       "S001                        11.52               10.23                 11.50                11.47  \n",
       "S002                        11.64               10.64                 11.26                11.57  \n",
       "S003                        11.60               10.21                 11.51                11.09  \n",
       "S005                        11.90               10.21                 11.34                11.51  \n",
       "S006                        11.97                9.77                 11.37                12.35  \n",
       "\n",
       "[5 rows x 28083 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_and_tran = en.join_metadata_to_omics(metadata_df_name=\"clinical\", omics_df_name=\"transcriptomics\")\n",
    "clin_and_tran.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Joining only specific columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Age</th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>ZZZ3_transcriptomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>64.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>58.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>50.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>75.0</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>11.51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>63.0</td>\n",
       "      <td>Serous</td>\n",
       "      <td>12.35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Age Histologic_type  ZZZ3_transcriptomics\n",
       "Sample_ID                                            \n",
       "S001       64.0    Endometrioid                 11.47\n",
       "S002       58.0    Endometrioid                 11.57\n",
       "S003       50.0    Endometrioid                 11.09\n",
       "S005       75.0    Endometrioid                 11.51\n",
       "S006       63.0          Serous                 12.35"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clin_and_tran = en.join_metadata_to_omics(\n",
    "    metadata_df_name=\"clinical\", \n",
    "    omics_df_name=\"transcriptomics\", \n",
    "    metadata_cols = [\"Age\", \"Histologic_type\"], \n",
    "    omics_genes=\"ZZZ3\")\n",
    "\n",
    "clin_and_tran.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_metadata_to_metadata</code>\n",
    "\n",
    "The <code>join_metadata_to_metadata</code> function joins two metadata dataframes (e.g. clinical or derived_molecular) to each other. Note how we passed a column name to select from the clinical dataframe, but passing <code>None</code> for the column parameter for the derived_molecular dataframe caused the entire dataframe to be selected. We could have omitted the <code>cols2</code> parameter altogether, as it is assigned to None by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>Estrogen_Receptor</th>\n",
       "      <th>Estrogen_Receptor_%</th>\n",
       "      <th>Progesterone_Receptor</th>\n",
       "      <th>Progesterone_Receptor_%</th>\n",
       "      <th>MLH1</th>\n",
       "      <th>MLH2</th>\n",
       "      <th>MSH6</th>\n",
       "      <th>PMS2</th>\n",
       "      <th>p53</th>\n",
       "      <th>...</th>\n",
       "      <th>Log2_variant_total</th>\n",
       "      <th>Log2_SNP_total</th>\n",
       "      <th>Log2_INDEL_total</th>\n",
       "      <th>Genomics_subtype</th>\n",
       "      <th>Mutation_signature_C&gt;A</th>\n",
       "      <th>Mutation_signature_C&gt;G</th>\n",
       "      <th>Mutation_signature_C&gt;T</th>\n",
       "      <th>Mutation_signature_T&gt;C</th>\n",
       "      <th>Mutation_signature_T&gt;A</th>\n",
       "      <th>Mutation_signature_T&gt;G</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>10.062046</td>\n",
       "      <td>9.984418</td>\n",
       "      <td>5.832890</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>8.300395</td>\n",
       "      <td>1.482213</td>\n",
       "      <td>72.529644</td>\n",
       "      <td>14.426877</td>\n",
       "      <td>1.383399</td>\n",
       "      <td>1.877470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Loss of nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>8.861087</td>\n",
       "      <td>8.330917</td>\n",
       "      <td>7.169925</td>\n",
       "      <td>MSI-H</td>\n",
       "      <td>14.641745</td>\n",
       "      <td>2.803738</td>\n",
       "      <td>64.485981</td>\n",
       "      <td>15.264798</td>\n",
       "      <td>0.934579</td>\n",
       "      <td>1.869159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>5.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>3.169925</td>\n",
       "      <td>CNV_low</td>\n",
       "      <td>16.129032</td>\n",
       "      <td>3.225806</td>\n",
       "      <td>70.967742</td>\n",
       "      <td>3.225806</td>\n",
       "      <td>3.225806</td>\n",
       "      <td>3.225806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>...</td>\n",
       "      <td>5.672425</td>\n",
       "      <td>5.523562</td>\n",
       "      <td>2.584963</td>\n",
       "      <td>CNV_low</td>\n",
       "      <td>17.777778</td>\n",
       "      <td>8.888889</td>\n",
       "      <td>62.222222</td>\n",
       "      <td>8.888889</td>\n",
       "      <td>2.222222</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Intact nuclear expression</td>\n",
       "      <td>Normal</td>\n",
       "      <td>...</td>\n",
       "      <td>6.108524</td>\n",
       "      <td>5.954196</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>CNV_high</td>\n",
       "      <td>9.836066</td>\n",
       "      <td>13.114754</td>\n",
       "      <td>62.295082</td>\n",
       "      <td>3.278689</td>\n",
       "      <td>8.196721</td>\n",
       "      <td>3.278689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 126 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Histologic_type     Estrogen_Receptor  Estrogen_Receptor_% Progesterone_Receptor  Progesterone_Receptor_%                       MLH1                       MLH2                        MSH6  \\\n",
       "Sample_ID                                                                                                                                                                                               \n",
       "S001         Endometrioid  Cannot be determined                  NaN  Cannot be determined                      NaN  Intact nuclear expression  Intact nuclear expression  Loss of nuclear expression   \n",
       "S002         Endometrioid  Cannot be determined                  NaN  Cannot be determined                      NaN  Intact nuclear expression  Intact nuclear expression   Intact nuclear expression   \n",
       "S003         Endometrioid  Cannot be determined                  NaN  Cannot be determined                      NaN  Intact nuclear expression  Intact nuclear expression   Intact nuclear expression   \n",
       "S005         Endometrioid  Cannot be determined                  NaN  Cannot be determined                      NaN  Intact nuclear expression  Intact nuclear expression   Intact nuclear expression   \n",
       "S006               Serous  Cannot be determined                  NaN  Cannot be determined                      NaN  Intact nuclear expression  Intact nuclear expression   Intact nuclear expression   \n",
       "\n",
       "                                 PMS2                   p53  ... Log2_variant_total Log2_SNP_total  Log2_INDEL_total  Genomics_subtype  Mutation_signature_C>A  Mutation_signature_C>G  \\\n",
       "Sample_ID                                                    ...                                                                                                                         \n",
       "S001        Intact nuclear expression  Cannot be determined  ...          10.062046       9.984418          5.832890             MSI-H                8.300395                1.482213   \n",
       "S002       Loss of nuclear expression  Cannot be determined  ...           8.861087       8.330917          7.169925             MSI-H               14.641745                2.803738   \n",
       "S003        Intact nuclear expression  Cannot be determined  ...           5.321928       5.000000          3.169925           CNV_low               16.129032                3.225806   \n",
       "S005        Intact nuclear expression  Cannot be determined  ...           5.672425       5.523562          2.584963           CNV_low               17.777778                8.888889   \n",
       "S006        Intact nuclear expression                Normal  ...           6.108524       5.954196          3.000000          CNV_high                9.836066               13.114754   \n",
       "\n",
       "           Mutation_signature_C>T  Mutation_signature_T>C  Mutation_signature_T>A  Mutation_signature_T>G  \n",
       "Sample_ID                                                                                                  \n",
       "S001                    72.529644               14.426877                1.383399                1.877470  \n",
       "S002                    64.485981               15.264798                0.934579                1.869159  \n",
       "S003                    70.967742                3.225806                3.225806                3.225806  \n",
       "S005                    62.222222                8.888889                2.222222                0.000000  \n",
       "S006                    62.295082                3.278689                8.196721                3.278689  \n",
       "\n",
       "[5 rows x 126 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_and_derived_molecular = en.join_metadata_to_metadata(\n",
    "    df1_name=\"clinical\",\n",
    "    df2_name=\"derived_molecular\",\n",
    "    cols1=\"Histologic_type\") # Note that we can omit the cols2 parameter, and it will by default select all of df2.\n",
    "                             # We could have also omitted cols1, if we wanted to select all of df1.\n",
    "\n",
    "hist_and_derived_molecular.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_omics_to_mutations</code>\n",
    "\n",
    "The <code>join_omics_to_mutations</code> function joins an -omics dataframe with the mutation data for a specified gene or genes. Because there may be multiple mutations for one gene in a single sample, the mutation type and location data are returned in lists by default, even if there is only one mutation. If there is no mutation for the gene in a particular sample, the list contains either \"Wildtype_Tumor\" or \"Wildtype_Normal\", depending on whether it's a tumor or normal sample. The mutation status column contains either \"Single_mutation\", \"Multiple_mutation\", \"Wildtype_Tumor\", or \"Wildtype_Normal\", for help with parsing.\n",
    "\n",
    "(Note: You can hide the Location columns by passing <code>False</code> to the optional <code>show_location</code> parameter.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AURKA_proteomics</th>\n",
       "      <th>TP53_proteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2950</td>\n",
       "      <td>[Missense_Mutation, Nonsense_Mutation]</td>\n",
       "      <td>[p.R130Q, p.R233*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>0.31100</td>\n",
       "      <td>0.2770</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.G127R]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.8710</td>\n",
       "      <td>[Nonsense_Mutation]</td>\n",
       "      <td>[p.W111*]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>-0.79800</td>\n",
       "      <td>-0.3430</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R130G]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>3.11000</td>\n",
       "      <td>3.0100</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S007</th>\n",
       "      <td>-1.65000</td>\n",
       "      <td>-0.1480</td>\n",
       "      <td>[Missense_Mutation, Missense_Mutation]</td>\n",
       "      <td>[p.Y68C, p.R130G]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4410</td>\n",
       "      <td>[Frame_Shift_Ins, Nonsense_Mutation]</td>\n",
       "      <td>[p.H118Qfs*8, p.Y180*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S009</th>\n",
       "      <td>0.84800</td>\n",
       "      <td>-1.2200</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S010</th>\n",
       "      <td>-1.73000</td>\n",
       "      <td>-0.0825</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R130G]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S011</th>\n",
       "      <td>-0.00513</td>\n",
       "      <td>-0.1810</td>\n",
       "      <td>[Missense_Mutation, Frame_Shift_Ins]</td>\n",
       "      <td>[p.H93R, p.E242*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AURKA_proteomics  TP53_proteomics                           PTEN_Mutation           PTEN_Location PTEN_Mutation_Status Sample_Status\n",
       "Sample_ID                                                                                                                                      \n",
       "S001                    NaN           0.2950  [Missense_Mutation, Nonsense_Mutation]      [p.R130Q, p.R233*]    Multiple_mutation         Tumor\n",
       "S002                0.31100           0.2770                     [Missense_Mutation]               [p.G127R]      Single_mutation         Tumor\n",
       "S003                    NaN          -0.8710                     [Nonsense_Mutation]               [p.W111*]      Single_mutation         Tumor\n",
       "S005               -0.79800          -0.3430                     [Missense_Mutation]               [p.R130G]      Single_mutation         Tumor\n",
       "S006                3.11000           3.0100                        [Wildtype_Tumor]           [No_mutation]       Wildtype_Tumor         Tumor\n",
       "S007               -1.65000          -0.1480  [Missense_Mutation, Missense_Mutation]       [p.Y68C, p.R130G]    Multiple_mutation         Tumor\n",
       "S008                    NaN           0.4410    [Frame_Shift_Ins, Nonsense_Mutation]  [p.H118Qfs*8, p.Y180*]    Multiple_mutation         Tumor\n",
       "S009                0.84800          -1.2200                        [Wildtype_Tumor]           [No_mutation]       Wildtype_Tumor         Tumor\n",
       "S010               -1.73000          -0.0825                     [Missense_Mutation]               [p.R130G]      Single_mutation         Tumor\n",
       "S011               -0.00513          -0.1810    [Missense_Mutation, Frame_Shift_Ins]       [p.H93R, p.E242*]    Multiple_mutation         Tumor"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_acet_and_PTEN_mut = en.join_omics_to_mutations(\n",
    "    omics_df_name=\"proteomics\",\n",
    "    mutations_genes=\"PTEN\", \n",
    "    omics_genes=[\"AURKA\", \"TP53\"])\n",
    "\n",
    "selected_acet_and_PTEN_mut.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering multiple mutations\n",
    "\n",
    "The function has the ability to filter multiple mutations down to just one mutation. It allows you to specify particular mutation types or locations to prioritize, and also provides a default sorting hierarchy for all other mutations. The default hierarchy chooses truncation mutations over missense mutations, and silent mutations last of all. If there are multiple mutations of the same type, it chooses the mutation occurring earlier in the sequence. \n",
    "\n",
    "To filter all mutations based on this default hierarchy, simply pass an empty list to the optional <code>mutations_filter</code> parameter. Notice how in sample S001, the nonsense mutation was chosen over the missense mutation, because it's a type of trucation mutation, even though the missense mutation occurs earlier in the peptide sequence. In sample S008, both mutations were types of truncation mutations, so the function just chose the earlier one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AURKA_proteomics</th>\n",
       "      <th>TP53_proteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.R233*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.441</td>\n",
       "      <td>Frame_Shift_Ins</td>\n",
       "      <td>p.H118Qfs*8</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AURKA_proteomics  TP53_proteomics      PTEN_Mutation PTEN_Location PTEN_Mutation_Status Sample_Status\n",
       "Sample_ID                                                                                                       \n",
       "S001                    NaN            0.295  Nonsense_Mutation       p.R233*    Multiple_mutation         Tumor\n",
       "S008                    NaN            0.441    Frame_Shift_Ins   p.H118Qfs*8    Multiple_mutation         Tumor"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PTEN_default_filter = en.join_omics_to_mutations(omics_df_name=\"proteomics\", mutations_genes=\"PTEN\", \n",
    "                                                        omics_genes=[\"AURKA\", \"TP53\"],\n",
    "                                                        mutations_filter=[])\n",
    "PTEN_default_filter.loc[[\"S001\", \"S008\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To prioritize a particular type of mutation, or a particular location, include it in the <code>mutations_filter</code> list. Below, we tell the function to prioritize nonsense mutations over all other mutations. Notice how in sample S008, the nonsense mutation is now selected instead of the frameshift insertion, even though the nonsense mutation occurs later in the peptide sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AURKA_proteomics</th>\n",
       "      <th>TP53_proteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.R233*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.441</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.Y180*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AURKA_proteomics  TP53_proteomics      PTEN_Mutation PTEN_Location PTEN_Mutation_Status Sample_Status\n",
       "Sample_ID                                                                                                       \n",
       "S001                    NaN            0.295  Nonsense_Mutation       p.R233*    Multiple_mutation         Tumor\n",
       "S008                    NaN            0.441  Nonsense_Mutation       p.Y180*    Multiple_mutation         Tumor"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PTEN_simple_filter = en.join_omics_to_mutations(omics_df_name=\"proteomics\", mutations_genes=\"PTEN\", \n",
    "                                                        omics_genes=[\"AURKA\", \"TP53\"], \n",
    "                                                        mutations_filter=[\"Nonsense_Mutation\"])\n",
    "PTEN_simple_filter.loc[[\"S001\", \"S008\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can include multiple mutation types and/or locations in the <code>mutations_filter</code> list. Values earlier in the list will be prioritized over values later in the list. For example, with the filter we specify below, the function first selects sample S001's missense mutation over its nonsense mutation, because we put the location of S001's missense mutation as the first value in our filter list. We still included Nonsense_Mutation in the filter list, but it comes after the location of S001's missense mutation, which is why S001's missense mutation is still prioritized. However, on all other samples, unless they also have a mutation at that same location, the function will continue prioritizing nonsense mutations, as we see in sample S008."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AURKA_proteomics</th>\n",
       "      <th>TP53_proteomics</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130Q</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S008</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.441</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.Y180*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           AURKA_proteomics  TP53_proteomics      PTEN_Mutation PTEN_Location PTEN_Mutation_Status Sample_Status\n",
       "Sample_ID                                                                                                       \n",
       "S001                    NaN            0.295  Missense_Mutation       p.R130Q    Multiple_mutation         Tumor\n",
       "S008                    NaN            0.441  Nonsense_Mutation       p.Y180*    Multiple_mutation         Tumor"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PTEN_complex_filter = en.join_omics_to_mutations(omics_df_name=\"proteomics\", mutations_genes=\"PTEN\", \n",
    "                                                        omics_genes=[\"AURKA\", \"TP53\"], \n",
    "                                                        mutations_filter=[\"p.R130Q\", \"Nonsense_Mutation\"])\n",
    "PTEN_complex_filter.loc[[\"S001\", \"S008\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <code>join_metadata_to_mutations</code>\n",
    "\n",
    "The <code>join_metadata_to_mutations</code> function works exactly like <code>join_omics_to_mutations</code>, except that it works with metadata dataframes (e.g. clinical and derived molecular) instead of omics dataframes. It also can filter multiple mutations, which you control through the <code>mutations_filter</code> parameter, and has the ability to hide the location colunms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Missense_Mutation, Nonsense_Mutation]</td>\n",
       "      <td>[p.R130Q, p.R233*]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.G127R]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Nonsense_Mutation]</td>\n",
       "      <td>[p.W111*]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R130G]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>Serous</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Histologic_type                           PTEN_Mutation       PTEN_Location PTEN_Mutation_Status Sample_Status\n",
       "Sample_ID                                                                                                               \n",
       "S001         Endometrioid  [Missense_Mutation, Nonsense_Mutation]  [p.R130Q, p.R233*]    Multiple_mutation         Tumor\n",
       "S002         Endometrioid                     [Missense_Mutation]           [p.G127R]      Single_mutation         Tumor\n",
       "S003         Endometrioid                     [Nonsense_Mutation]           [p.W111*]      Single_mutation         Tumor\n",
       "S005         Endometrioid                     [Missense_Mutation]           [p.R130G]      Single_mutation         Tumor\n",
       "S006               Serous                        [Wildtype_Tumor]       [No_mutation]       Wildtype_Tumor         Tumor"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_and_PTEN = en.join_metadata_to_mutations(\n",
    "    metadata_df_name=\"clinical\",\n",
    "    mutations_genes=\"PTEN\",\n",
    "    metadata_cols=\"Histologic_type\")\n",
    "\n",
    "hist_and_PTEN.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With multiple mutations filtered:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Histologic_type</th>\n",
       "      <th>PTEN_Mutation</th>\n",
       "      <th>PTEN_Location</th>\n",
       "      <th>PTEN_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.R233*</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.G127R</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S003</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>p.W111*</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S005</th>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R130G</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S006</th>\n",
       "      <td>Serous</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>No_mutation</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Histologic_type      PTEN_Mutation PTEN_Location PTEN_Mutation_Status Sample_Status\n",
       "Sample_ID                                                                                    \n",
       "S001         Endometrioid  Nonsense_Mutation       p.R233*    Multiple_mutation         Tumor\n",
       "S002         Endometrioid  Missense_Mutation       p.G127R      Single_mutation         Tumor\n",
       "S003         Endometrioid  Nonsense_Mutation       p.W111*      Single_mutation         Tumor\n",
       "S005         Endometrioid  Missense_Mutation       p.R130G      Single_mutation         Tumor\n",
       "S006               Serous     Wildtype_Tumor   No_mutation       Wildtype_Tumor         Tumor"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist_and_PTEN = en.join_metadata_to_mutations(\n",
    "    metadata_df_name=\"clinical\",\n",
    "    mutations_genes=\"PTEN\",\n",
    "    metadata_cols=\"Histologic_type\",\n",
    "    mutations_filter=[\"Nonsense_Mutation\"])\n",
    "\n",
    "hist_and_PTEN.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting dataframes\n",
    "\n",
    "If you wish to export a dataframe to a file, simply call the dataframe's to_csv method, passing the path you wish to save the file to, and the value separator you want:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist_and_PTEN.to_csv(path_or_buf=\"histologic_type_and_PTEN_mutation.tsv\", sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
