{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use Case for TP53 Downstream Effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to show more of what we can do with our mutation data, let's look at the gene TP53. TP53 is a tumor suppressing gene that is very important in helping prevent mutation. Attached [here](https://en.wikipedia.org/wiki/P53) is a wikipedia article for a further introduction of TP53."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we know, a mutation in a gene often causes significant funcional changes to the protein that it produces. We know that sometimes an alteration to that specific protein can cause immediate harmful effects to the cell. But in order to measure the mutation's complete effect, we also need to look at the proteins that interact with our mutated protein directly, as well as the ones that may be involved later on in a process that involves many proteins. This use case will show how to find some of those proteins, and how a mutation in our gene (TP53) affects them as well."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Import Libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will import standard python libraries used to analyze and visualize data as pandas dataframes, as well as our CPTAC data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                    \r"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import cptac\n",
    "import cptac.utils as al\n",
    "\n",
    "en = cptac.Endometrial()\n",
    "co = cptac.Colon()\n",
    "ccrcc = cptac.Ccrcc()\n",
    "\n",
    "gene = \"TP53\"\n",
    "\n",
    "omics = \"proteomics\"\n",
    "#omics = \"transcriptomics\"\n",
    "#omics = \"phosphoproteomics\"\n",
    "#omics = \"acetylproteomics\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Get Interacting Proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will get a list of the proteins that are know to interact with TP53 directly. The method identified here, the bioplex method, is a process that includes tagging our protein of interest and then identifying the proteins around it that come into direct contact. When we call the get_interacting_proteins_bioplex method, this returns a list of proteins that interact with our gene in this way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interacting Proteins:\n",
      "HNF4A\n",
      "FBXW7\n",
      "MDM4\n",
      "CELA2B\n",
      "FAM111A\n",
      "MDM2\n",
      "RORC\n",
      "AURKA\n",
      "FAM175B\n",
      "GPR156\n",
      "HOMER3\n",
      "CUL7\n",
      "CUL9\n"
     ]
    }
   ],
   "source": [
    "interacting_proteins = al.get_interacting_proteins_bioplex(gene)\n",
    "\n",
    "print(\"Interacting Proteins:\")\n",
    "for interacting_protein in interacting_proteins:\n",
    "    print(interacting_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3: Run T-tests to Find Significant Comparisons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we create a new dataframe containing protein measurements for the interacting proteins that we are interested in. This is done with endometrial cancer, because we know that mutations in TP53 have been shown to be especially important in endometrial cancer, but can also be done across all of the other cancers in the dataset. Later on in this use case we will show an example with colon cancr."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: The following columns were not found in the proteomics dataframe, so they were inserted into joined table, but filled with NaN: CELA2B, FAM175B, FBXW7, GPR156, HNF4A, MDM2, MDM4, RORC (/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 1)\n",
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 123 samples for the TP53 gene (/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 1)\n"
     ]
    }
   ],
   "source": [
    "protdf = en.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics, omics_genes=interacting_proteins)\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, since in order to make comparisons we will want our \"Label\" column to contain one of two values- either \"Mutated\" or \"Wildtype.\" We will then delete other columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>HNF4A_proteomics</th>\n",
       "      <th>FBXW7_proteomics</th>\n",
       "      <th>MDM4_proteomics</th>\n",
       "      <th>CELA2B_proteomics</th>\n",
       "      <th>FAM111A_proteomics</th>\n",
       "      <th>MDM2_proteomics</th>\n",
       "      <th>RORC_proteomics</th>\n",
       "      <th>AURKA_proteomics</th>\n",
       "      <th>FAM175B_proteomics</th>\n",
       "      <th>GPR156_proteomics</th>\n",
       "      <th>HOMER3_proteomics</th>\n",
       "      <th>CUL7_proteomics</th>\n",
       "      <th>CUL9_proteomics</th>\n",
       "      <th>TP53_Mutation</th>\n",
       "      <th>TP53_Location</th>\n",
       "      <th>TP53_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>S001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.363000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.443</td>\n",
       "      <td>0.0408</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.R248W]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.601000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.897</td>\n",
       "      <td>-0.1430</td>\n",
       "      <td>0.1480</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.280000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.451</td>\n",
       "      <td>0.0999</td>\n",
       "      <td>0.1130</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.066500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.798</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.034</td>\n",
       "      <td>-0.7460</td>\n",
       "      <td>0.1680</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.000411</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.110</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.535</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>-0.2530</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.S241C]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       HNF4A_proteomics  FBXW7_proteomics  MDM4_proteomics  \\\n",
       "Sample_ID                                                        \n",
       "S001                    NaN               NaN              NaN   \n",
       "S002                    NaN               NaN              NaN   \n",
       "S003                    NaN               NaN              NaN   \n",
       "S005                    NaN               NaN              NaN   \n",
       "S006                    NaN               NaN              NaN   \n",
       "\n",
       "Name       CELA2B_proteomics  FAM111A_proteomics  MDM2_proteomics  \\\n",
       "Sample_ID                                                           \n",
       "S001                     NaN            0.363000              NaN   \n",
       "S002                     NaN            0.601000              NaN   \n",
       "S003                     NaN           -0.280000              NaN   \n",
       "S005                     NaN            0.066500              NaN   \n",
       "S006                     NaN           -0.000411              NaN   \n",
       "\n",
       "Name       RORC_proteomics  AURKA_proteomics  FAM175B_proteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                   NaN               NaN                 NaN   \n",
       "S002                   NaN             0.311                 NaN   \n",
       "S003                   NaN               NaN                 NaN   \n",
       "S005                   NaN            -0.798                 NaN   \n",
       "S006                   NaN             3.110                 NaN   \n",
       "\n",
       "Name       GPR156_proteomics  HOMER3_proteomics  CUL7_proteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                     NaN             -0.443           0.0408   \n",
       "S002                     NaN             -0.897          -0.1430   \n",
       "S003                     NaN             -0.451           0.0999   \n",
       "S005                     NaN             -0.034          -0.7460   \n",
       "S006                     NaN             -0.535          -0.1200   \n",
       "\n",
       "Name       CUL9_proteomics        TP53_Mutation  TP53_Location  \\\n",
       "Sample_ID                                                        \n",
       "S001                0.0355  [Missense_Mutation]      [p.R248W]   \n",
       "S002                0.1480     [Wildtype_Tumor]  [No_mutation]   \n",
       "S003                0.1130     [Wildtype_Tumor]  [No_mutation]   \n",
       "S005                0.1680     [Wildtype_Tumor]  [No_mutation]   \n",
       "S006               -0.2530  [Missense_Mutation]      [p.S241C]   \n",
       "\n",
       "Name      TP53_Mutation_Status Sample_Status     Label  \n",
       "Sample_ID                                               \n",
       "S001           Single_mutation         Tumor   Mutated  \n",
       "S002            Wildtype_Tumor         Tumor  Wildtype  \n",
       "S003            Wildtype_Tumor         Tumor  Wildtype  \n",
       "S005            Wildtype_Tumor         Tumor  Wildtype  \n",
       "S006           Single_mutation         Tumor   Mutated  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we are only interested in interacting protein levels, here we trim down the dataframe and drop unwanted columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take the column names, which are the interacting proteins, and create a list in order to run our t-tests for each protein individually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_list = list(protdf.columns)\n",
    "col_list.remove('Label')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HNF4A_proteomics', 'FBXW7_proteomics', 'MDM4_proteomics', 'CELA2B_proteomics', 'FAM111A_proteomics', 'MDM2_proteomics', 'RORC_proteomics', 'AURKA_proteomics', 'FAM175B_proteomics', 'GPR156_proteomics', 'HOMER3_proteomics', 'CUL7_proteomics', 'CUL9_proteomics']\n"
     ]
    }
   ],
   "source": [
    "print(col_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we call wrap_ttest, which is an accessory function in cptac.utils. It expects one column with a binary variable (which is 'Label' for us as either mutated or wildtype), and all other data as qualitiative. It will then run ttests to determine if there is a statistically significant difference between the binary values and their respective qualitative data. For our case, it will tell us if there are significantly different measurements between wildtype and mutated samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n",
      "1 significant comparisons!\n",
      "         Comparison   P_Value\n",
      "0  AURKA_proteomics  0.000016\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Doing t-test comparisons\\n\")\n",
    "wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "if wrap_results is not None:\n",
    "        print(wrap_results)\n",
    "        print(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our t-test returned one protein whose levels were significantly changed by a mutation in TP53, aurora kinase. Attached [here](https://www.ncbi.nlm.nih.gov/pubmed/27933271) is an article that talks about the relationship between TP53 and aurora kinase. We will now create a boxplot to show exactly how levels of aurora kinase change with our gene's mutation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 4: Analyze Selected Protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have selected the protein that we want to analyze, we will now show how to visualize the effects that a mutation in TP53 had on aurora kinase. First we want to create and trim down a dataframe in order to easily graph our results. To do this, we simply want the sample ID, the protein level, and a label as to whether the sample is mutated or wildtype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 123 samples for the TP53 gene (/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>AURKA_proteomics</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>S001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S002</td>\n",
       "      <td>0.311</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S005</td>\n",
       "      <td>-0.798</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S006</td>\n",
       "      <td>3.110</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       AURKA_proteomics     Label\n",
       "Sample_ID                            \n",
       "S001                    NaN   Mutated\n",
       "S002                  0.311  Wildtype\n",
       "S003                    NaN  Wildtype\n",
       "S005                 -0.798  Wildtype\n",
       "S006                  3.110   Mutated"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protdf = en.join_omics_to_mutations(omics_df_name='proteomics',omics_genes='AURKA',mutations_genes='TP53')\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "\n",
    "'''Create the binary valued column needed to do the comparison'''\n",
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "'''Format the dataframe correctly'''\n",
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "protdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will plot our data using the seaborn library as a stripplot on top of a boxplot in order to see the distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x1a38f75f98>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAEGCAYAAACD7ClEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3deXxU9dX48c+ZmewLIQtbkEXDWpFi44JQlwesqNWqj7ZFWar+lLY+ap9qW6u2VtuXrVbpD6ld7AO/Yi0ufYpKEdnUIq2CghsugAFZEllCErJvM3N+f8wkZLKQCczkJpnzfr14mXvn3jsnOMy53+Wer6gqxhhjYpPL6QCMMcY4x5KAMcbEMEsCxhgTwywJGGNMDLMkYIwxMczjdABdlZ2drSNGjHA6DGOM6VW2bNlyWFVzWu/vdUlgxIgRbN682ekwjDGmVxGRPe3tt+4gY4yJYZYEjDEmhlkSMMaYGGZJwBhjYpglAWOMiWGWBIwxJoZZEjDGmBhmSSCGlZSUcNttt1FSUuJ0KMYYh1gS6OV2797Ngw8+yJ133smKFSu6dO6SJUvYunUrTz75ZJSiM8b0dL3uiWFzVH19PT/5yU8oLy8HYMeOHSQlJTFt2rROzy0pKWHVqlWoKqtWrWLOnDlkZWVFO2RjTA9jLYFebNu2bc0JoMmmTZvCOnfJkiX4/X4AfD6ftQaMiVGWBHqxwYMH43KF/i8cMmRIWOeuW7cOr9cLgNfrZe3atRGPzxjT8zmaBEQkUUTeEpH3ReQjEbnfyXh6mwEDBjB79mzi4uIAGD16NFdddVVY506fPh2PJ9Ab6PF4uPDCC6MWpzGm5xInF5oXEQFSVLVKROKAfwG3q+rGjs7Jz89XqyIaqrKyksrKyrBbARAYE5g5cyYNDQ0kJCSwdOlSGxMwpg8TkS2qmt96v6MtAQ2oCm7GBf84l5V6qbS0tC4lAICsrCxmzJiBiDBjxgxLAMbEKMfHBETELSLvAYeAtaraZmRTRG4Wkc0isrm4uLj7g+yj5s6dy4QJE5gzZ47ToRhjHOJod1BLIpIBPA/cqqofdnScdQcZY0zX9cjuoJZU9QjwT2CGw6EYY0zMcHp2UE6wBYCIJAHTgW1OxmSMMbHE6SeGBwNLRMRNICE9p6pdq31gjDHmuDmaBFT1A2CSkzEYY0ws6zFjAsYYY7qfJQFjjIlhlgSMMSaGWRIwxpgYZknAGGNimCUBY4yJYZYEjDEmhlkSMMaYGGZJwBhjYpglAWOMiWGWBIwxJoZZEjDGmBhmScAYY2KYJQFjjIlhlgSMMSaGWRIwxpgYZknAGGNimCUBY4yJYZYEjDEmhlkSMMaYGGZJwBhjYpglAWOMiWGWBIwxJoZZEjDGmBhmScAYY2KYJYEYp6rs27ePmpoap0MxxjjA43QAxjkHDhzggQceoLCwkMTERObNm8e0adOcDssY041EVZ2OoUvy8/N18+bNTodxQhYuXEhBQYHTYbBnzx78fn/ztogwbNgwXK7ubSDm5eVx6623dut7GhNrRGSLqua33m/dQTHM5/OFbKtqm33GmL7NuoMc0FPueufMmcORI0eat4cNG8bChQsREQejMsZ0J0sCMSwjIwMRIScnhyFDhjBr1ixLAMbEGEeTgIicBDwJDAL8wBOqusDJmGKJiJCRkcEjjzzidCjGGIc43RLwAneo6jsikgZsEZG1qvqxw3EZY0xMcHRgWFX3q+o7wZ8rgU+AXCdjMsaYWNJjZgeJyAhgErCpndduFpHNIrK5uLi4u0Mzxpg+q0ckARFJBf4OfE9VK1q/rqpPqGq+qubn5OR0f4DGGNNHOZ4ERCSOQAL4q6ouczoeY4yJJY4mAQnMR1wEfKKq852MxRhjYpHTLYEpwGzgP0TkveCfSxyOyRhjYoajU0RV9V+APZ1kjDEOcbolYIwxxkFhJQERmSIiKcGfZ4nIfBEZHt3QjDHGRFu4LYHfAzUiMhH4IbCHQLkH08f5/X7eeustVqxYwYEDB5wOxxgTYeGOCXhVVUXka8ACVV0kInOjGZjpGR599FE2bNgAwJ///GceeOABxo8f73BUxphICbclUCkiPwZmAS+JiBuIi15YpifYv39/cwIAaGho4IUXXnAwImNMpIWbBL4B1AM3quoBAvV9fh21qEyP0N4CM16v14FIjDHREm4SSAJ+r6pNt4XFwOvRCcn0FEOHDuX0009v3na5XFx22WUORmSMibRwxwT+BpzTYtsX3HdGxCMyjjtw4ADLly+npqaGq666iilTpnDw4EEmT57MKaec4nR4xpgICjcJeFS1oWlDVRtEJD5KMRkH1dTU8MMf/rB52cn169fz0EMPceGFFzocmTEmGsLtDioWkcubNoKzhA5HJyTjpC1btoSsO+zz+fjnP//pXEDGmKgKtyXwbeCvIvJbAmUe9gFzohaVcUy/fv3C2meM6RvCagmo6k5VPRsYD4xX1XNUtSC6oRknTJgwgbPPPrt5e+jQoVx88cUORmSMiaZjtgREZJaqPiUi32+1HwAr/9z3iAh33303O3bsoKamhgkTJuB2u50OyxgTJZ11B6UE/5sW7UBMzzJ69GinQzDGdINjJgFV/WPwv/d3TzjGGGO6U1gDwyIyErgVGNHyHFW9vKNzeqqFCxdSUGDDGUDz38Ptt9/ucCQ9Q15eHrfeeqvTYRjTrcKdHfQCgWUg/wH4oxdO9BUUFPDeh5/gS850OhTHuRoUgC27DjocifPcNaVOh2CMI8JNAnWq+lhUI+lGvuRMasfaKpbmqKRtK50OwRhHhJsEFojIfcAaAoXkAFDVd6ISlTHGmG4RbhKYQHBBeI52B2lw2/RRnvpyPI3V1CfnoC6rHG5MXxRuErgSOLll/SDTt6Uf+oC0sh0A+NzxHD7pPLwJ9uSwMX1NuLWD3gcyohmI6TlcjbWkBhMAgNvXQFrJNgcjMsZES7gtgYHANhF5m9AxgV43RdR0zuVvQFrv89W3e6wxpncLNwncF9UoTI/iTehHQ2J/4uvKmvfV9BvhXEDGmKgJKwmo6noRGcjRRWTeUtVD0QvLOO3w0C+TWvYpnsZqalOHUpc2xOmQTAx77rnnePnll0lMTGTmzJmce+65TofUZ4T7xPDXCawp/E8CpaQXisgPVPV/oxib6QbJ5Z+RVFGILy6Zyqyx+OIC5aLUHU9l9hccjs4YeOONN3jqqaeat+fPn09eXh5DhtiNSSSE2x10D3BG092/iOQA6wBLAr1Ycvln9D+wpXk7oeYQB0fOAGk9ImBiRU8pq1JUVARAbm4uJSUlIa/5/X7uvfde0tKiX9cyFkqJhDs7yNWq+6ekC+eaHiqpojBk29NYTVydlU8wzqutraW2thaA+Pi2K9kmJCSgqtTW1lJdXY3f36ur2Tgq3JbAKhFZDTwd3P4G8HJ0QjLdxReXHLKtCD5PcgdHm1jQU+56m4oaLliwAL/fz5/+9CfWrFlDfHw8M2fO5LLLLuOBBx5gy5ZAS3bAgAE8/PDDZGZaTbCuCndg+AcichUwlcCYwBOq+nxUIzNRV5k1loSaYjyNVShCRfZ4/HFJTodlTAiXy8W8efO48cYbERHcbjcffPBBcwIAOHToECtXrmTWrFkORto7hdsSANgCVKjqOhFJFpE0Va2MVmAm+nxxKRwceRFxdWX44pLweywBmO5VU1PD7373OzZt2sTAgQO57bbb2ixo5PV6Wbx4Ma+99hoZGRnccMMNNDY2trlWVVVVd4Xdp4TVry8iNxEYBP5jcFcugfLSJ0xEFovIIRH5MBLXM10kQmNSpiUA44gnn3yS119/nfr6evbu3ctdd91FWVlZyDErVqxgxYoVVFdXU1RUxEMPPcSoUaPIzs5uPsbj8TBt2rTuDr9PCLclcAtwJrAJQFU/FZEBEYrhz8BvgScjdD1jTC/x/vvvh2x7vV6WLl0asu/jjz8O2W5oaGDfvn38+te/ZuXKlVRXVzNt2jRGjRoV9Xj7onCTQL2qNjQtMC8iHgJVRE+Yqr4uIiMicS1jTO8yePDg5umgTVq3BMaMGcPGjRubtz0eDyeffDIZGRnMnj27W+Lsy8Kd5rleRO4GkkTkQuBvBFYZ6xYicrOIbBaRzcXFxd31tsaYKPvud7+LxxN6Lzp58uSQ7csvv5xp06bh8XjIzs7m+9//PhkZVs8yUsJNAncBxcBWYB6wUlXviVpUrajqE6qar6r5OTk53fW2xpgoy87O5tFHH+Xss89mzJgxzJs3r03f/vbt2ykvL2f8+PHMmzePqVOnOhRt3xRud9CtqroA+FPTDhG5PbjP9GSqgILYs32m52hsbOSNN96grKyMc845h7vvvrvd4w4ePMh9993XPBvoww8/5OGHH24zg8gcv3CTwFyg9Rf+t9rZZ7qBp76c1NIduPxeqjNOpj5lYLvHJVXso9+h93H56qlNG0rZoHxwubs5WmNCqSo/+9nP2Lp1KwBLly7loYceYuTIkW2O3bJlS8h0UL/fz5tvvhmSBPx+P2vXruXjjz9mzJgxXHTRRbjd9jkP1zGTgIjMBK4FRorI8hYvpRMoHXHCRORp4HwgW0QKgftUdVEkrt0Xia+BnL3/xOUP/MNIrCri8LDzaUjKDjnO5a2j/4G3EQ08Tp9cuQ9vQjqVWeO6PWZjWtq+fXtzAgCoq6tjxYoV7T6tPGjQoDb7Bg8eHLK9ZMkSnn8+8Ozqa6+9xt69e/n2t78d4aj7rs5aAm8A+4Fs4NEW+yuBDyIRgKrOjMR1wlVUVIS7ppykbSu7820jJh4vLjl6ZyRA2p5/UUNofRUPPkRC66kkFm/DW/xZd4TZ67hrSigq8jodRkxor85PR7V/Jk2axPTp03nllVdQVc4880wuuOCCkGPWrFkTsr1u3TrmzZuHWCHEsBwzCajqHmAPMLnVegKfqKr9i3GAv82aX+3v8+FCNbQgaKPV/DM9wLhx4xg3bhyffPIJECgQd+mll7Z7rIhw2223MXPmTHw+X7stg7S0NKqrq5u3U1NTLQF0QbjrCVwDPEIfWE8gNzeXA/Ueasde4nQox0cVz/63SK7cB0BDQgZHTjoPdce1OdRXfZD04g9we+uoTR9Gec4EGyDuQNK2leTmtj+2YiJLRHjggQfYsGEDZWVlTJkypdO1AY41K3DOnDk8+uij+Hw+XC4Xc+fOjXTIfVq4A8P3YusJ9AwilA05i8r6cYi/kcbEzA7r/9enDKQ45cJuDtCYziUkJDB9+vSIXGvq1KmMGTOG7du3M2rUKAYOtGTeFeEmAVtPoIfxJqR3+Jr4Gsk4uIWkqs9pjE+jfODpNCRldWN0xnSvnJycY7YWTMfC/SJfJSKrReRbIvIt4CWgd46sxoD0wx+SXFmIqJ/4+nIyP38T1BbdMMa0ZesJ9EHxtYdDtt3eOjyN1Xjjo78cnzGmd+k0CYiIG1itqtOBZdEPyZyohsQs4uvLm7d97kS8wQXkjTGmpU67g1TVB9SISL9uiMdEQEXOqdSm5qIIjfHplA4522YFGWPaFe7AcB2wVUTWAs0TclX1tqhEZU6IuuMpzZ3c+YHGmJgXbhJ4KfjHGGPC9umnn1JWVsbEiRNJSEhwOhzTjnAHhpeISDwwlsBiMttVtSGqkZkeK662lMSqz/HFp1KTPsy6mky7FixYwCuvvAJAZmYmv/rVr9p94rer/H4/r776Ktu2bWPMmDFMmzYNl8s+g8cr3CeGLyGwvvBOArODRorIPFV9OZrBmcgTvxcVd4cPmHUmsepzMoveaC5UkVhVRGnulMgFaPqE3bt3NycAgNLSUl544YVjFnbz+Xzs2rWLAQOOvXJty4Jxa9asYd++fdxwww2RCTwGhdsdNB+4QFULAETkFALdQ5YEegnxNZK5fxMJ1QfwuxM4MnASdWlDu3ydlLKCkEpFSVX7cTdU4YtPjVywpterrKwMa1+TwsJC7rvvPoqLi/F4PPTr14/09PYfiFy1alXI9urVqy0JnIBw21CHmhJA0C7gUEcHm54nreRjEqsPIIDbV0///W8jvuPo0WvV9aPt7DNm/Pjx5ObmNm+LSJsVw1p66qmnaFo61uv1Ulpa2mFl0ZSUlGNum64JtyXwkYisBJ4j8O/+GuDt4ANkqKo9P9DDxdcdCdl2qQ9PQyWNXSwnUZk5moSaQ83rFNSkD8cXlxyxOE3f4Ha7efDBB1mxYgWlpaWcf/75TJw4scPjDx0KvadUVXw+X7vHzp49mwULFuD3+3G5XMyaNSuisceacJNAInAQOC+4XQxkApcRSAq9Kgm4a0p77XoCx8tHIy37cfwK8dvXEg/4E482uwUlkUbc+GnETT0eaFWqupx44vDhx0Vj+SGSynv/36W7phSwwmOR1L9/f2bPnh3WsVOnTqWg4GhnQ3x8PHFxbSvjAlxwwQVUVlby4osvkpCQYIPCJyjc2UHXH+t1Efmxqv4yMiFFV15entMhOEJVKS0tpbq6Go/HQ2ZmJlpYCEDeyUe//Pbv309dXWCpiHj85Gak0L9/f0di7l4DY/az0RNceeWVxMXFsXHjRnJzc/noo486PHbnzp0sWrQIVQXgN7/5DUOGDLF1h49TuC2BzlwD9Iok0N4SdrHq9ttvBwJT+QDKy8vb3LmlpKQ0v25MOOrr69m5cycHDx5k8+bNZGZmcsUVV5CVFdr1WFRUxLPPPktZWRnjx4/ntNNO49JLL8XlcjV/Ntvz7rvvNicACNzgvPPOO5YEjlOkkoAt49MHJCYm4vF48HqPLhqXnZ19jDOMCfXpp59y//33U1FREbJ/06ZN/P73v29eAL6xsZF77rmH0tJSAN5//32efvppRo4cyYMPPtjutVWVjz76KOTz2WTEiBGR/UViSKSSgHZ+iOnpNm7c2OYfWKQW/jCdW7hwYUi/eG8U6E6sa7P/wIEDfOc73yEpKQmA2tra5gTQ0meffcYtt9zSPFOoqUVQV1fHwYMHm2cMuVyu5p/T0tJYunQpTz/9dFR+p54gLy8var0Y1hIwzbZv395mX319vQORxKaCggI+/ehdhqW2PyumN2j0ZwDudl/zHfyEegl8cfvVBWTQ3ldHfWkR8Y2BRFK/ZzNedXFIQ4/1+/2kUU2K1OOqLqGhendkf5EeZG9V+3+fkXLcSUBEzlDVt4Obf4tQPMZB48ePZ8WKFc3bIsK4ceMcjCj2DEv1cffpFZ0f2EO9VOTi5f1tJxKckVnJ3JMD05QP1nnYUZHE7upK3ipJQ1t8uQvKd8eWMjz16DMs/zyYzv/ua5ssJg9q5PKh5W329zUPvtPxKoKR0KUkICLjgW8CM4FyIB9AVdvvxDO9ytSpU9m7dy8vvfQSiYmJXHvttZx00klOh2V6kYuHHKGgMpFPq5JC9p+cGmhRflCWzP/sHIA/+MV//oBy9tXEszN4vCJsLU8OSQL949uOAbjwc0ZWVbR+jZgSzqIywwl86c8EvMBwIF9Vd0c3NOOEa6+9lmuvvdbpMEwv5RI4K7uqTRIYnNQIwKr9Gc0JAOBfxWn4NPQuf+PhNL6ae4Ran7C8MJNPKxPJjG+ktCHw3EC/OC//5+SDzdc0J+aYSUBE3gD6Ac8AV6vqpyLymSUAY0xHzsiqYntFEptLU3AJXDCgnLy0QB9/Y6svfL8KiS4/tf6j/d5pcYExkWf2ZLOl9GhNqrzUGq4eVsbQZCtgHEmdPWpXDKQReJQyJ7jPZgIZYzrkFph7cjG/nLiXn526l9zkBj4oS8bnh/MHhI53nJlVxZUnlSLBr5U48XNZbhkAHx4JLUeysyqJIUmWACLtmC0BVf1acFnJ/wTuF5E8IENEzlTVt7olQmNMr+RHmL99CGUNga+ZkSl1fG/sfrISGvmkPIkhSY2ckVWFS2Bseh2FNfGMTK0jLS4wg2hgYiN7a44uRDMgsRGXzUOMuE7HBFS1HFgMLBaRAcA3gP8rIiepqo0aGmPa9e/itOYEAPBZdSIflSdzWkYNY9NDnyXITPCSmRA6AHz1SYd5Yucgqrxu0uO8zBx+uFvijjXH7A4SkZCyf6p6SFUXquo59JIyEcYYZzT42962f1qRCEC9T9p9vUlVo4u/7smhyhsYKxieXM8pqfbMSjR01hJ4XkSuUdUtLXeKyM+Ay4HfRSswY0zvdlZWFesPpdPgP3qv+dqhfhysi2NbZRJuUaYPLOfS3CNtzt1QnM7Buvjm7a3lKeyoTGRMetunkc2J6Wxg+BrgbyIyGUAC/gCcC5wfiQBEZIaIbBeRAhG5KxLXNMY4b1BSI1cOLWmz/+OKZPwqNPoDD5btrGy7AH2Vt+1XU1OrwERWZwPDW0TkCgItgluAm4IvzYjEQvMi4gYeBy4ECgksVLNcVT8+0Wsb09sUFRVRXemO+hOi3alRO//iXrwjnRQJ7eppUCUwETHQZeTCz4u7EnhuVxpu8ZNMXfMgca3GU6HJ+BGSqSddao53Ce0eaU+lm5Sioqhdv7MxgUwCX85zgaeARmAekBp87USdCRSo6q5gUnkG+FoErmuM6QHixEcyx+rCUeK17UNf8eIjSypIoo4UakmjmlLSqSaJCk2hVAOJ0qcuyjQVH24UF9UkUUPbloXpWGdjAls4+lxAJXAW8BaB9KzAySf4/rnAvhbbhcH3MCbm5ObmUu/d36trB7WvgkN1Hsob3SzcPjjkiWEQbvpCNUM6efr3jwUD2Hrk6HkNxDFnbB2H6jws3hV62+9zJ1EtiXyxfzVXDC0lrpcvPPbgO+kktFivOdI66w4aGbV3Dmiv0dbmYTQRuRm4GWDYsGFRDskYE2kDEr0MSPSSl1bHjsqjJSXcKOmezqumxrvaPqMa5/IzIqUBFxqSWJrGDtYf6keS289X2xl4NkcdV44UkTEi8qcIvH8h0PJZg6HA560PUtUnVDVfVfNzcnJav2yM6SW+Ofxwc0E4QTlvYDmpwYfDjmXawHISXEePOyOzioGJgWcLrhtxmPQ4Lx5pmyi2VyS12WdCdVY76DTgEWAI8AKwkMC00LOARyPw/m8Do0RkJFBEoEKpVS8zpo8akOjlppMPsmDHYOr9Ll49mIFH4PKhZcc8b1hKA/dNKOTDI0n0j/eGPGx2VnYVZ2VXUecV7t06jDrf0XtbqzPUuc5aAn8ClhIoG1EMvAPsAvJU9Tcn+uaq6gX+C1gNfAI8p6odrzBtjOn11hzMoL7FswPrDvSjojHQhbOvJp4PypJ5tzSZJz/LZkVRRvN00fQ4H+fkVDGuX127s38SPcqcEcWkxwVaGmPTa7k099jJxXQ+MJygqn8O/rxdRO4E7lLViC19pKorgZWRup4xpmeragy99/Qj1HhdrPo8g9eL206P3XokmbvGfx7WtM/T+tdwakYNDX4h0W21LsPRWRJIFJFJHB3ArQJOEwn871DVd6IZnDGm7zk7u4qCFusNjEypw+PSdhMAQFFtAvtq4hmWEl7XjkuwBNAFnSWBA8D8DrYV+I9oBGVOTEVFBQkJCSQkRH6+dE1NDStXruTAgQNMnjyZL33pSxF/D9O3nZ1dRZLbz3tlKeQkNnL+gAqK6zv+KhKUZE/ng8fm+HQ2RfT8borDREB9fT2PPPIImzZtIjExkeuuu46vfS2yz979/Oc/56OPAsM2a9as4Y477uC8886L6HuYvm9i/xom9q9p3h7mbmBoUj2FtW1vXM4dUEF2QtslJk1kdDY76KpWuxQ4DLynqpVRi6qPW7hwIQUFBRG/7pEjRygrCwyE1dXVsWjRIlavXk1cXFy7xzfFcPvtt4d1/YaGBopaPb7++OOPs2zZshOIGvLy8rj11ltP6BqmdxOB/xpzgOf3ZbKlNAWvuoh3+bhiaBnnDrCvmmjqrDvosnb2ZRIYF7hRVV+NQkzmODU0tO0zbWho6DAJJCV1bQ61y9V2Mll7+4w5HqkeP7NHHuabw0s4UBvHgMRGEqxvP+o66w66vr39wcXnn8NKPByXaN31vvbaa/zmN0dn7iYmJjJ//nzS0yNXkGzRokW8+OKLACQnJ/OLX/yCvLy8iF3fmDiXclKLQeCyBjdpHh87KpN45UA/AP5jYDlfyKh1KsQ+pdOVxdqjqntEpP3bS+OYCy64gCNHjrBu3TrS0tKYNWtWRBMAwI033sh5553HgQMHmDhxImlpaRG9vjFNius8PLFzIPtr40l2+6jzuZrLQ3xamciPxheRm3zsmkOmc8eVBERkLGDL/PRAV155JVdeeWVU3yMvL8/u/k2Xrd3fj9eL00lw+blkyBFOz6xufm3j4VRe+rw/jX7hyzkVXJp7hGWFmeyvDSwsU+MLLUntR9h6JIXcZKsLdKI6Gxj+B20LumUCg4FZ0QrKGNO3vF+WzItFR6vP/79dOQxNrmdAopf9tXH8dXc2GrzLf3l/f4YkNTQngI7kJForIBI6awk80mpbgVICiWAW8GY0gjImVu2t6luLygD4VSjW0N9JERZ81J8UqadaE5oTQJPFO/oBPvAc7XUW/MHjhCTqWfWZm9W7+9bfVXv2VrkZFcXrdzYwvL7pZxH5IoHibl8HPgP+HsW4jIk5fbWLrbi4GF9VVZv9KYNHk5CQgDQ0UN5q6rHPFY+q0j89ndraWuLi4sjMzCRYrACP57h6snulUUT3s9FZd9BoApU9ZwIlwLOAqOoFUYvImBjVV5+V+M53vkNVqyRwzTXXMHv27ObtlStXsnTpUhoaGrj44ot59913EREWLFjQpfcqLy9ny5YtDBgwgFNPPTUi8fd1naXTbcAG4DJVLQAQkf+OelTGmD5j7NixIQ8ZZmVlce21oRXjL7nkEi655BJUFRHhnXe6XpZs586d3H333dTWBqaOTps2LewHIWNZZ0/6/CeBekGvicifRGQa7a8GZowx7br++us588wzcblcDB8+nB/96Ee43e0vQF9bW8svfvEL9uzZw969e3n11fCfR122bFlzAgB45ZVX+PzzNmtUmVY6GxN4HnheRFKAK4D/BgaKyO+B51V1TTfEaIzpxdLT07n33nvDOnbZsmW89dZbAPj9fn77298yadIk+vfv3+m5dXVtF7Svr7eZ7J0J65l/Va1W1b+q6lcJLAH5HnBXVCMzxsScXbt2hWx7vV727dsX1rkzZsxoHjiGwJ7bvIMAAAznSURBVBPz69evt0TQiS4XflHVUlX9o6paGWljTERNnDgxZDslJYVRo8KbIHnGGWfwy1/+kuzsbCDQMli2bBmLFi2KeJx9iVX/Msb0GF/96le5+uqrcbvdJCQk8JOf/KRLhQ5PPvlkDh8+HLJv48aNkQ6zT7EkYIzpMdxuN3PmzGHYsGEMGTKE8ePHd+n8+Pj45pZAkyFDhkQyxD7HkoAxJiq8Xi/r16/n2WefZc+ePd3yni6Xi1tuuYXU1FQAsrOzuemmm7rlvXur2HnszhjTrR5++OHmrphnnnmGn/70p0yaNCmi76Gq+P1+Ghsbefzxx3nzzTcZOHAgd9xxBzk5OeTm5nY4HdUEWBIwxkTc559/HtIX7/P5WL58eZeSgNfrpaKiosNy6GvWrOHJJ5+kpqaGoUOHsnv3bgD27dvH/PnzWbx4sSWAMFh3kDEm4tpbca7l9M1jaWho4ODBg+zbt4+5c+eyePHiNsfs37+fxx9/nIqKCrxeb3MCaFJZWUlhYeFxxR5rLAkYYyJu0KBBTJkypXnb4/FwxRVXhHXuunXrqKkJLELv8/l44YUX2LZtGwBVVVUcOHCAgoICVDteejI1NZWhQ4eewG8QO6w7yBgTFXfeeSdf/vKX2bVrF42NjezcuZNhw4aRkZFxzPPau4MvLCzkk08+4amnnqKxsZFTTjkFt9uNz+drPmb06NHs2rWLQYMGMW/ePBISEiL+O/VFlgSMMVHhdrsZN24cf/jDHzhyJLAC2PLly3nssceOuSzpGWecwYoVK5q34+PjGTp0KHfddRd+vx8IFIs7++yzKSwspLKykosuuojrrrsu7C4nc5QlAWNM1Kxfv745AQCUlJSwYcMGLrnkkg7PmTRpEtnZ2VRWVjJ27Fi+/vWvU19f35wAWvrd734XlbhjiSUBY0zUtDc7J5wFYdLS0khLS+PnP/85ECgE169fP8rLy5uPOeussyIXaAyzgWFjTNScd955DBgwoHl78ODBTJ06tcvXSUhI4IEHHuCss85i1KhR3HTTTUybNi2SocYsawkYY6ImPT2dBQsW8O9//xsRYcqUKSQnJx/XtUaOHMk999wT4QiNJQFjTFSlpKTwla98xekwTAcc6w4SkWtE5CMR8YtIvlNxGGNMLHNyTOBD4CrgdQdjMMaYmOZYd5CqfgLhP0pujDEm8nrF7CARuVlENovI5uLiYqfDMcaYPiOqLQERWQcMauele1T1xXCvo6pPAE8A5Ofnd1wwxBhjTJdENQmo6vRoXt8YY8yJsSmixhhHFRQU8MQTT1BUVMRZZ53FzTff7HRIMcWxJCAiVwILgRzgJRF5T1UvcioeY0z38/l8PPjgg82Lw69bt46UlBSHo4otcqya3D1Rfn6+bt682ekwjOmTFi5cSEFBQbe9X0NDA0VFRSH74uPjqa2tBSAvL6/bYmlPXl4et956q6MxRIqIbFHVNs9kWXeQMcYxHo8Hl8sVUiHU1gHoXpYEjDHNnLjrfffdd3n88ccpLi5m0qRJfP/73+9wXWETedYdZIxxnKri9XqJi4tzOpQ+q6PuoF7xsJgxpm8TEUsADrEkYIwxMcySgDHGxDBLAsYYE8MsCRhjTAyzJGCMMTHMkoAxxsQwSwLGGBPDLAkYY0wMsyRgjDExzJKAMcbEMEsCxhgTwywJGGNMDLMkYIzpcUpKSrjtttsoKSlxOpQ+z5KAMabHWbJkCVu3buXJJ5/s8rnl5eVs376dxsbGKETW91gSMMb0KCUlJaxatQpVZdWqVV1qDaxevZrrr7+eH/zgB9x8883s2bMnipH2DZYEjDE9ypIlS/D5fAB4vd6wWwN1dXUsWrQIr9cLBJLJX/7yl6jF2VdYEjDGOMrv94d03axbt645Cfh8PtauXRvWdSorK6mrqwvZd+jQocgF2kdZEjDGOOa1115j7ty5XHPNNTz66KM0NDQwderUkGO+/OUvh3WtnJwcRo8eHbKv9bVMW7bQvDHGESUlJTz22GPNd/3r169n+PDhiMhxX/Pee+/l2WefpbCwkPz8fC677LJIhdtnWRIwxjjis88+a04ATQoKCtiwYUPIvg0bNvDjH/84rGtmZGQwb968iMUYC6w7yBjjiNGjRxMfHx+yb8KECUyfPh2PJ3B/6vF4uPDCC50IL2ZYEjDGOCI9PZ27776bkSNHkpGRwVVXXcWMGTOYO3cuLlfgq8ntdjNnzhyHI+3brDvIGOOY008/ndNPPz1kX1ZWFjNmzOAf//gHM2bMICsry6HoYoMlAWNMjzN37lx2795trYBuYEnAGNPjZGVl8dhjjzkdRkywMQFjjIlhlgSMMSaGOZYEROTXIrJNRD4QkedFJMOpWIwxJlY52RJYC5yqqqcBO4DwngYxxhgTMY4lAVVdo6re4OZGYKhTsRhjTKzqKWMCNwAvd/SiiNwsIptFZHNxcXE3hmWMMX1bVKeIisg6YFA7L92jqi8Gj7kH8AJ/7eg6qvoE8ARAfn6+RiFUY4yJSaLq3HeqiMwFvg1MU9WaMM8pBmy5oMjJBg47HYQx7bDPZmQNV9Wc1jsdSwIiMgOYD5ynqtbH4xAR2ayq+U7HYUxr9tnsHk6OCfwWSAPWish7IvIHB2MxxpiY5FjZCFXNc+q9jTHGBPSU2UHGOU84HYAxHbDPZjdwdGDYGGOMs6wlYIwxMcySgDHGxDBLAr2MiKiI/KXFtkdEikVkRSfnZYjId8O4fljHtXPez0Tkzq6eZ3o/EfmNiHyvxfZqEfmfFtuPisjdIvK/we3zO/q8ishuEck+3s+h6TpLAr1PNXCqiCQFty8EisI4LwMI5x9VuMcZ0+QN4BwAEXEReMjrCy1ePwd4RVWv7sI17XPYTSwJ9E4vA5cGf54JPN30Qus7chH5UERGAL8CTgk+k/FrEUkVkVdE5B0R2SoiXwueEnJc8Bo/EJG3g2W/729x7XtEZHuwPMiYaP7Cpkf7N8EkQODL/0OgUkT6i0gCMA4oE5EPW58oIlkiskZE3hWRPwISfKn15/UvLT6jiMhfReRyEfmWiLwoIquCn8X7WhwzS0TeCl7jjyLijtLv36tZEuidngG+KSKJwGnApjDOuQvYqapfVNUfAHXAlap6OnAB8KiISOvjROQrwCjgTOCLwJdE5FwR+RLwTWAScBVwRoR/R9NLqOrngFdEhhFIBm8S+ExOBvKBD4CGDk6/D/iXqk4ClgPDgvtbf17/B7geQET6Bd9nZfDYM4HrCHw+rxGRfBEZB3wDmKKqXwR8wWNMK7bGcC+kqh8E7+5ncvQfQlcJ8KCInAv4gVxgYDvHfSX4593gdiqBpJAGPN9U80lElh9nHKZvaGoNnEOgHExu8OdyAt1FHTmXwE0EqvqSiJS1d5CqrheRx0VkQPD4v6uqN3DfwlpVLQEQkWXAVAJFKb8EvB08Jgk4dKK/ZF9kSaD3Wg48ApwPZLXY7yW0hZfYwfnXATnAl1S1UUR2d3CsAL9U1T+G7AwMBNpDJqZJ07jABALdQfuAO4AKYHEn54b7OfoLgc/tNwmUn+/ofCXwuV2iqrZYVSesO6j3Wgw8oKpbW+3fDZwOICKnAyOD+ysJ3L036QccCiaAC4DhHRy3GrhBRFKD18wN3o29DlwpIkkikgZcFrHfzPRG/wa+CpSqqk9VSwkM7k4m0D3UkdcJdtOIyMVA/+D+1p9DgD8D3wNQ1Y9a7L9QRDKDkyWuCMbyCnB18LNK8PXhmDasJdBLqWohsKCdl/4OzBGR94C3CSzdiaqWiMi/g4NzLwMPAf8Qkc3Ae8C29o4LjguMA94MNqurgFmq+o6IPBs8dw+wIZq/r+nxthKYFbS01b5UVT3cdBPRjvuBp0XkHWA9sBfa/xyq6kER+QR4odU1/kWglZAHLFXVzQAici+wJjhjqRG4BStD34aVjTDG9AoikkwgsZyuquXBfd8C8lX1v5yMrTez7iBjTI8nItMJtFYXNiUAExnWEjDGmBhmLQFjjIlhlgSMMSaGWRIwxpgYZknAmA6ISFUXju1yFdWuXN+YaLEkYIwxMcySgDFdICKXicimYNXLdSLSst7SRBF5VUQ+FZGbWpzTbhVWY3oCSwLGdM2/gLODVS+fAX7Y4rXTCJT4ngz8VESGdFSFtZtjNqZDVjbCmK4ZCjwrIoOBeOCzFq+9qKq1QK2IvEbgi38q7Vdhfb37QjamY5YEjOmahcB8VV0uIucDP2vxWkfVLNtUYTWmp7DuIGO6ph9Hl/Oc2+q1r4lIoohkESjx/TYdV2E1pkewloAxHUsWkcIW2/MJ3Pn/TUSKgI0cLdUN8BbwEoHVsX4eXHHr8/aqsGILnJgewmoHGWNMDLPuIGOMiWGWBIwxJoZZEjDGmBhmScAYY2KYJQFjjIlhlgSMMSaGWRIwxpgY9v8BGL9t4scmBoAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.boxplot(data = protdf, x = 'Label', y = 'AURKA_proteomics')\n",
    "sns.stripplot(data = protdf, x = 'Label', y = 'AURKA_proteomics', jitter = True, dodge = True, color = '.3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 5: Example Using General Transcription Factor (ARID1A)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have analyzed proteins that interact with TP53, we will now look at a general transcription factor, ARID1A. Proteins made by this gene interact with a wide range of genes, as it alters chromatin structure around them. Attached [here](https://en.wikipedia.org/wiki/ARID1A) is a wikipedia article about ARID1A for further reading."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code box will go through thousands of proteins contained in our dataset and look at how a mutation in ARID1A affects them. Because of the high amount of comparisons, this will take a little bit of time. In the end, we will have a list of genes affected by ARID1A."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene:  ARID1A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 101 samples for the ARID1A gene (/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/numpy/core/fromnumeric.py:3367: RuntimeWarning: Degrees of freedom <= 0 for slice\n",
      "  **kwargs)\n",
      "/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/numpy/core/_methods.py:132: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 significant comparisons!\n",
      "          Comparison       P_Value\n",
      "0  ARID1A_proteomics  1.025018e-10\n",
      "1   MAP1S_proteomics  1.010149e-07\n",
      "2    DPF2_proteomics  2.166609e-07\n",
      "3   KDM4B_proteomics  4.414547e-07\n",
      "4    RRM1_proteomics  4.455870e-07\n",
      "5  MAP7D3_proteomics  4.880890e-07\n",
      "6   CYTH3_proteomics  3.109396e-06\n",
      "7   CEP78_proteomics  3.749172e-06\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "gene = \"ARID1A\"\n",
    "print(\"\\nGene: \", gene)\n",
    "\n",
    "'''Use all proteins'''\n",
    "\n",
    "'''Create dataframe in order to do comparisons with wrap_ttest'''\n",
    "protdf = en.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics) #Here is the different line\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "    \n",
    "'''Create the binary valued column needed to do the comparison'''\n",
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "'''Format the dataframe correctly'''\n",
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "'''Make list of columns to be compared using t-tests'''\n",
    "col_list = list(protdf.columns)\n",
    "col_list.remove('Label')\n",
    "\n",
    "print(\"Doing t-test comparisons\\n\")\n",
    "    \n",
    "'''Call wrap_ttest, pass in formatted dataframe'''\n",
    "wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "'''Print results, if anything significant was found'''\n",
    "if wrap_results is not None:\n",
    "        print(wrap_results)\n",
    "        print(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each of the proteins shown can be individually analyzed in the same manner that we showed above. For further ideas about how to analyze larger lists of genes, refer to usecase 5."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 6: Use the same tools on another cancer type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have looked at endometrial cancer, we will now use the same tools on colon cancer. We will find that there will not be any similar results. This goes to show how unique cancer types are, as around 20 percent of our colon cancer samples have mutations in ARID1A, but there are no similar effects like we see in endometrial cancer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene:  ARID1A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 184 samples for the ARID1A gene (/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 5)\n",
      "Warning: proteomics data was not found for the following samples, so proteomics data columns were filled with NaN for these samples: S001, S013, S015, S017, S038, S080, S081, S090, S099, S109 (/Users/kimball.benjamin1/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 5)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n",
      "No significant comparisons.\n"
     ]
    }
   ],
   "source": [
    "print(\"\\nGene: \", gene)\n",
    "    \n",
    "'''Use all proteins'''\n",
    "'''Create dataframe in order to do comparisons with wrap_ttest'''\n",
    "protdf = co.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics)\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "\n",
    "'''Create the binary valued column needed to do the comparison'''\n",
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "'''Format the dataframe correctly'''\n",
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "'''Make list of columns to be compared using t-tests'''\n",
    "col_list = list(protdf.columns)\n",
    "col_list.remove('Label')\n",
    "\n",
    "print(\"Doing t-test comparisons\\n\")\n",
    "    \n",
    "'''Call wrap_ttest, pass in formatted dataframe'''\n",
    "wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "'''Print results, if anything significant was found'''\n",
    "if wrap_results is not None:\n",
    "        print(wrap_results)\n",
    "        print(\"\\n\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
