{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Use Case Trans Effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ARID1A Mutation Trans Effect on Proteomics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This use case will showcase the effects of ARID1A mutation, first on interacting proteins and then with a list of other proteins in endometrial and colon cancer. While we are only working with these two, the strategies shown here are easily applicable across the many other types of cancer data found in the CPTAC package."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Import Libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will import standard python libraries used to analyze and visualize data as pandas dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import CPTAC data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac\n",
    "import cptac.utils as al"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we define the datesets we will work with in this use case, endometrial and colon cancer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                    \r"
     ]
    }
   ],
   "source": [
    "en = cptac.Endometrial()\n",
    "co = cptac.Colon()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we identify the gene we will work with. For this case, we will use ARID1A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"ARID1A\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this use case we want to see how our gene's mutation affects proteins that interact with it, so we will use proteomics. You can easily decide to analyze transcriptomics, phosphoproteomics, or acetylproteomics by changing the value of \"omics.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "omics = \"proteomics\"\n",
    "#omics = \"transcriptomics\"\n",
    "#omics = \"phosphoproteomics\"\n",
    "#omics = \"acetylproteomics\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2: Set up Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we want to create a new dataframe that we will populate with proteins that statistically prove to be affected by the ARID1A mutation. We start by labeling the dataframe column names, getting it ready to add our list of proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_significant_comparisons = pd.DataFrame(columns=['Cancer_Type', 'Gene', 'Comparison','Interacting_Protein','P_Value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer_Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>Interacting_Protein</th>\n",
       "      <th>P_Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Cancer_Type, Gene, Comparison, Interacting_Protein, P_Value]\n",
       "Index: []"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_significant_comparisons.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we define a function that we will call to add the desired proteins and their corresponding relevant information to our dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_to_all_significant_comparisons(df, cancer, interacting, all_sig_comp):\n",
    "    expanded = df\n",
    "    expanded['Gene'] = gene\n",
    "    expanded['Cancer_Type'] = cancer\n",
    "    expanded['Interacting_Protein'] = interacting\n",
    "    \n",
    "    updated_all_comparisons = pd.concat([all_sig_comp, expanded], sort=False)\n",
    "    \n",
    "    return updated_all_comparisons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3: Get Interacting Proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will get a list of the proteins that are know to interact with ARID1A. We start with these because they have a high probability of being affected by a mutation because of their close contact. In this case we use the bioplex method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Interacting Proteins:\n",
      "SS18\n",
      "BCL7A\n",
      "SMARCE1\n",
      "DPF2\n",
      "SMARCC2\n",
      "BCL7C\n",
      "SMARCD1\n",
      "DPF3\n"
     ]
    }
   ],
   "source": [
    "interacting_proteins = al.get_interacting_proteins_bioplex(gene)\n",
    "\n",
    "print(\"Interacting Proteins:\")\n",
    "for interacting_protein in interacting_proteins:\n",
    "    print(interacting_protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test for significant comparisons in any of interacting proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3: Run T-tests to find significant comparisons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Endometrial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we create a new dataframe containing protein measurements for the interacting proteins that we are interested in. This is done with endometrial cancer, but as mentioned earlier can be done for other cancers. An example with colon cancer will be shown later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 101 samples for the ARID1A gene (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 1)\n"
     ]
    }
   ],
   "source": [
    "protdf = en.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics, omics_genes=interacting_proteins)\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, since in order to make comparisons we will want our \"Label\" column to contain one of two values- either \"Mutated\" or \"Wildtype.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>SS18_proteomics</th>\n",
       "      <th>BCL7A_proteomics</th>\n",
       "      <th>SMARCE1_proteomics</th>\n",
       "      <th>DPF2_proteomics</th>\n",
       "      <th>SMARCC2_proteomics</th>\n",
       "      <th>BCL7C_proteomics</th>\n",
       "      <th>SMARCD1_proteomics</th>\n",
       "      <th>DPF3_proteomics</th>\n",
       "      <th>ARID1A_Mutation</th>\n",
       "      <th>ARID1A_Location</th>\n",
       "      <th>ARID1A_Mutation_Status</th>\n",
       "      <th>Sample_Status</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>S001</td>\n",
       "      <td>-0.498</td>\n",
       "      <td>0.4100</td>\n",
       "      <td>-0.3950</td>\n",
       "      <td>-0.548</td>\n",
       "      <td>-0.451</td>\n",
       "      <td>-0.9180</td>\n",
       "      <td>-0.328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Missense_Mutation]</td>\n",
       "      <td>[p.T2121P]</td>\n",
       "      <td>Single_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S002</td>\n",
       "      <td>-0.162</td>\n",
       "      <td>0.8880</td>\n",
       "      <td>-0.5560</td>\n",
       "      <td>-0.919</td>\n",
       "      <td>-0.627</td>\n",
       "      <td>-0.9410</td>\n",
       "      <td>-0.418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Nonsense_Mutation, Frame_Shift_Del]</td>\n",
       "      <td>[p.Q403*, p.D1850Tfs*33]</td>\n",
       "      <td>Multiple_mutation</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S003</td>\n",
       "      <td>-0.448</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>-0.5400</td>\n",
       "      <td>-0.555</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>-0.3880</td>\n",
       "      <td>-0.648</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S005</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.6880</td>\n",
       "      <td>0.0968</td>\n",
       "      <td>0.372</td>\n",
       "      <td>-0.102</td>\n",
       "      <td>-0.0328</td>\n",
       "      <td>0.186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S006</td>\n",
       "      <td>0.298</td>\n",
       "      <td>0.3460</td>\n",
       "      <td>0.0897</td>\n",
       "      <td>-0.642</td>\n",
       "      <td>-0.432</td>\n",
       "      <td>-1.4000</td>\n",
       "      <td>-0.599</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[Wildtype_Tumor]</td>\n",
       "      <td>[No_mutation]</td>\n",
       "      <td>Wildtype_Tumor</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       SS18_proteomics  BCL7A_proteomics  SMARCE1_proteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                -0.498            0.4100             -0.3950   \n",
       "S002                -0.162            0.8880             -0.5560   \n",
       "S003                -0.448            0.0927             -0.5400   \n",
       "S005                 0.203            0.6880              0.0968   \n",
       "S006                 0.298            0.3460              0.0897   \n",
       "\n",
       "Name       DPF2_proteomics  SMARCC2_proteomics  BCL7C_proteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                -0.548              -0.451           -0.9180   \n",
       "S002                -0.919              -0.627           -0.9410   \n",
       "S003                -0.555              -0.537           -0.3880   \n",
       "S005                 0.372              -0.102           -0.0328   \n",
       "S006                -0.642              -0.432           -1.4000   \n",
       "\n",
       "Name       SMARCD1_proteomics  DPF3_proteomics  \\\n",
       "Sample_ID                                        \n",
       "S001                   -0.328              NaN   \n",
       "S002                   -0.418              NaN   \n",
       "S003                   -0.648              NaN   \n",
       "S005                    0.186              NaN   \n",
       "S006                   -0.599              NaN   \n",
       "\n",
       "Name                            ARID1A_Mutation           ARID1A_Location  \\\n",
       "Sample_ID                                                                   \n",
       "S001                        [Missense_Mutation]                [p.T2121P]   \n",
       "S002       [Nonsense_Mutation, Frame_Shift_Del]  [p.Q403*, p.D1850Tfs*33]   \n",
       "S003                           [Wildtype_Tumor]             [No_mutation]   \n",
       "S005                           [Wildtype_Tumor]             [No_mutation]   \n",
       "S006                           [Wildtype_Tumor]             [No_mutation]   \n",
       "\n",
       "Name      ARID1A_Mutation_Status Sample_Status     Label  \n",
       "Sample_ID                                                 \n",
       "S001             Single_mutation         Tumor   Mutated  \n",
       "S002           Multiple_mutation         Tumor   Mutated  \n",
       "S003              Wildtype_Tumor         Tumor  Wildtype  \n",
       "S005              Wildtype_Tumor         Tumor  Wildtype  \n",
       "S006              Wildtype_Tumor         Tumor  Wildtype  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we are only interested in interacting protein levels, here we trim down the dataframe and drop unwanted columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>SS18_proteomics</th>\n",
       "      <th>BCL7A_proteomics</th>\n",
       "      <th>SMARCE1_proteomics</th>\n",
       "      <th>DPF2_proteomics</th>\n",
       "      <th>SMARCC2_proteomics</th>\n",
       "      <th>BCL7C_proteomics</th>\n",
       "      <th>SMARCD1_proteomics</th>\n",
       "      <th>DPF3_proteomics</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>S001</td>\n",
       "      <td>-0.498</td>\n",
       "      <td>0.4100</td>\n",
       "      <td>-0.3950</td>\n",
       "      <td>-0.548</td>\n",
       "      <td>-0.451</td>\n",
       "      <td>-0.9180</td>\n",
       "      <td>-0.328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S002</td>\n",
       "      <td>-0.162</td>\n",
       "      <td>0.8880</td>\n",
       "      <td>-0.5560</td>\n",
       "      <td>-0.919</td>\n",
       "      <td>-0.627</td>\n",
       "      <td>-0.9410</td>\n",
       "      <td>-0.418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S003</td>\n",
       "      <td>-0.448</td>\n",
       "      <td>0.0927</td>\n",
       "      <td>-0.5400</td>\n",
       "      <td>-0.555</td>\n",
       "      <td>-0.537</td>\n",
       "      <td>-0.3880</td>\n",
       "      <td>-0.648</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S005</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.6880</td>\n",
       "      <td>0.0968</td>\n",
       "      <td>0.372</td>\n",
       "      <td>-0.102</td>\n",
       "      <td>-0.0328</td>\n",
       "      <td>0.186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S006</td>\n",
       "      <td>0.298</td>\n",
       "      <td>0.3460</td>\n",
       "      <td>0.0897</td>\n",
       "      <td>-0.642</td>\n",
       "      <td>-0.432</td>\n",
       "      <td>-1.4000</td>\n",
       "      <td>-0.599</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       SS18_proteomics  BCL7A_proteomics  SMARCE1_proteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                -0.498            0.4100             -0.3950   \n",
       "S002                -0.162            0.8880             -0.5560   \n",
       "S003                -0.448            0.0927             -0.5400   \n",
       "S005                 0.203            0.6880              0.0968   \n",
       "S006                 0.298            0.3460              0.0897   \n",
       "\n",
       "Name       DPF2_proteomics  SMARCC2_proteomics  BCL7C_proteomics  \\\n",
       "Sample_ID                                                          \n",
       "S001                -0.548              -0.451           -0.9180   \n",
       "S002                -0.919              -0.627           -0.9410   \n",
       "S003                -0.555              -0.537           -0.3880   \n",
       "S005                 0.372              -0.102           -0.0328   \n",
       "S006                -0.642              -0.432           -1.4000   \n",
       "\n",
       "Name       SMARCD1_proteomics  DPF3_proteomics     Label  \n",
       "Sample_ID                                                 \n",
       "S001                   -0.328              NaN   Mutated  \n",
       "S002                   -0.418              NaN   Mutated  \n",
       "S003                   -0.648              NaN  Wildtype  \n",
       "S005                    0.186              NaN  Wildtype  \n",
       "S006                   -0.599              NaN  Wildtype  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take the column names, which are the interacting proteins, and create a list in order to run our t-tests for each protein individually."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_list = list(protdf.columns)\n",
    "col_list.remove('Label')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SS18_proteomics', 'BCL7A_proteomics', 'SMARCE1_proteomics', 'DPF2_proteomics', 'SMARCC2_proteomics', 'BCL7C_proteomics', 'SMARCD1_proteomics', 'DPF3_proteomics']\n"
     ]
    }
   ],
   "source": [
    "print(col_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we run our t-test algorithm from cptac.utils which will return to us a dataframe with any comparisons values that reach an adjusted threshold. These proteins are the ones that are significantly affected by the ARID1A mutation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n",
      "           Comparison       P_Value\n",
      "0     DPF2_proteomics  2.166609e-07\n",
      "1  SMARCC2_proteomics  8.270473e-05\n",
      "2  SMARCE1_proteomics  1.498027e-04\n",
      "3  SMARCD1_proteomics  1.744915e-04\n",
      "4    BCL7C_proteomics  4.379537e-04\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Doing t-test comparisons\\n\")\n",
    "\n",
    "wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "if wrap_results is not None:\n",
    "        print(wrap_results)\n",
    "        print(\"\\n\\n\")\n",
    "        \n",
    "        all_significant_comparisons = add_to_all_significant_comparisons(wrap_results, \"Endometrial\", True, all_significant_comparisons)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Colon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we follow the same steps that we took for endometrial cancer, but this time with colon cancer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: The following columns were not found in the proteomics dataframe, so they were inserted into joined table, but filled with NaN: BCL7A, DPF3, SS18 (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 2)\n",
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 184 samples for the ARID1A gene (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 2)\n",
      "Warning: proteomics data was not found for the following samples, so proteomics data columns were filled with NaN for these samples: S001, S013, S015, S017, S038, S080, S081, S090, S099, S109 (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 2)"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n",
      "No significant comparisons.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "'''Create dataframe in order to do comparisons with wrap_ttest'''\n",
    "protdf = co.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics, omics_genes=interacting_proteins)\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "\n",
    "'''Create the binary valued column needed to do the comparison'''\n",
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "'''Format the dataframe correctly'''\n",
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "'''Make list of columns to be compared using t-tests'''\n",
    "col_list = list(protdf.columns)\n",
    "col_list.remove('Label')\n",
    "\n",
    "print(\"Doing t-test comparisons\\n\")\n",
    "\n",
    "'''Call wrap_ttest, pass in formatted dataframe'''\n",
    "wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "'''Print results, if anything significant was found'''\n",
    "if wrap_results is not None:\n",
    "        print(wrap_results)\n",
    "        print(\"\\n\\n\")\n",
    "        \n",
    "        all_significant_comparisons = add_to_all_significant_comparisons(wrap_results, \"Colon\", True, all_significant_comparisons)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since no significant comparisons were found in colon cancer, no proteins or values were added to our overall dataframe."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 4: Find significant comparisons among all proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Endometrial"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have found significant values among interacting proteins, we will now go through the rest of the proteins in the CPTAC dataset and find any more that were significantly changed because of a mutation in ARID1A. The reason for doing two separate comparisons is because of the threshold used by wrap_ttest. The value is adjusted to be smaller and smaller as more proteins in a set are analyzed, as we use the formula of (.05 / number of proteins). Here we will once again go through the same t-testing steps from earlier, but this time without specifying specific proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene:  ARID1A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 101 samples for the ARID1A gene (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 7)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n",
      "          Comparison       P_Value\n",
      "0  ARID1A_proteomics  1.025018e-10\n",
      "1   MAP1S_proteomics  1.010149e-07\n",
      "2    DPF2_proteomics  2.166609e-07\n",
      "3   KDM4B_proteomics  4.414547e-07\n",
      "4    RRM1_proteomics  4.455870e-07\n",
      "5  MAP7D3_proteomics  4.880890e-07\n",
      "6   CYTH3_proteomics  3.109396e-06\n",
      "7   CEP78_proteomics  3.749172e-06\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(\"\\nGene: \", gene)\n",
    "\n",
    "    '''Use all proteins'''\n",
    "\n",
    "    '''Create dataframe in order to do comparisons with wrap_ttest'''\n",
    "    protdf = en.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics) #Here is the different line\n",
    "    protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "    \n",
    "    '''Create the binary valued column needed to do the comparison'''\n",
    "    for ind, row in protdf.iterrows():\n",
    "        if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "            protdf.at[ind,'Label'] = 'Mutated'\n",
    "        else:\n",
    "            protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "    '''Format the dataframe correctly'''\n",
    "    protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "    protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "    protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "    protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "    '''Make list of columns to be compared using t-tests'''\n",
    "    col_list = list(protdf.columns)\n",
    "    col_list.remove('Label')\n",
    "\n",
    "    print(\"Doing t-test comparisons\\n\")\n",
    "    \n",
    "    '''Call wrap_ttest, pass in formatted dataframe'''\n",
    "    wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "    '''Print results, if anything significant was found'''\n",
    "    if wrap_results is not None:\n",
    "            print(wrap_results)\n",
    "            print(\"\\n\\n\")\n",
    "            \n",
    "            all_significant_comparisons = add_to_all_significant_comparisons(wrap_results, \"Endometrial\", False, all_significant_comparisons)\n",
    "\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"Error in Comparison\")\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Colon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we do the same thing for colon samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Gene:  ARID1A\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 184 samples for the ARID1A gene (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 6)\n",
      "Warning: proteomics data was not found for the following samples, so proteomics data columns were filled with NaN for these samples: S001, S013, S015, S017, S038, S080, S081, S090, S099, S109 (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 6)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Doing t-test comparisons\n",
      "\n",
      "No significant comparisons.\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    print(\"\\nGene: \", gene)\n",
    "    \n",
    "    '''Use all proteins'''\n",
    "    '''Create dataframe in order to do comparisons with wrap_ttest'''\n",
    "    protdf = co.join_omics_to_mutations(mutations_genes=[gene], omics_df_name=omics)\n",
    "    protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "    \n",
    "    '''Create the binary valued column needed to do the comparison'''\n",
    "    for ind, row in protdf.iterrows():\n",
    "        if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "            protdf.at[ind,'Label'] = 'Mutated'\n",
    "        else:\n",
    "            protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "    '''Format the dataframe correctly'''\n",
    "    protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "    protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "    protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "    protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "    '''Make list of columns to be compared using t-tests'''\n",
    "    col_list = list(protdf.columns)\n",
    "    col_list.remove('Label')\n",
    "\n",
    "    print(\"Doing t-test comparisons\\n\")\n",
    "    \n",
    "    '''Call wrap_ttest, pass in formatted dataframe'''\n",
    "    wrap_results = al.wrap_ttest(protdf, 'Label', col_list)\n",
    "\n",
    "    '''Print results, if anything significant was found'''\n",
    "    if wrap_results is not None:\n",
    "            print(wrap_results)\n",
    "            print(\"\\n\\n\")\n",
    "            \n",
    "            all_significant_comparisons = add_to_all_significant_comparisons(wrap_results, \"Colon\", False, all_significant_comparisons)\n",
    "\n",
    "\n",
    "except Exception as e:\n",
    "    print(\"Error in Comparison\")\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 5: Analyze collected proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we will print our dataframe containing the affected proteins that we found. Note that ARID1A_proteomics has the smallest P-Value and is therefore the most affected, as expected since we are analyzing a mutation in ARID1A itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer_Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Comparison</th>\n",
       "      <th>Interacting_Protein</th>\n",
       "      <th>P_Value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>DPF2_proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>2.166609e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SMARCC2_proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>8.270473e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SMARCE1_proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>1.498027e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>SMARCD1_proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>1.744915e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>BCL7C_proteomics</td>\n",
       "      <td>True</td>\n",
       "      <td>4.379537e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>ARID1A_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>1.025018e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>MAP1S_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>1.010149e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>DPF2_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>2.166609e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>KDM4B_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>4.414547e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>RRM1_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>4.455870e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>MAP7D3_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>4.880890e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>CYTH3_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>3.109396e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>Endometrial</td>\n",
       "      <td>ARID1A</td>\n",
       "      <td>CEP78_proteomics</td>\n",
       "      <td>False</td>\n",
       "      <td>3.749172e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cancer_Type    Gene          Comparison Interacting_Protein       P_Value\n",
       "0  Endometrial  ARID1A     DPF2_proteomics                True  2.166609e-07\n",
       "1  Endometrial  ARID1A  SMARCC2_proteomics                True  8.270473e-05\n",
       "2  Endometrial  ARID1A  SMARCE1_proteomics                True  1.498027e-04\n",
       "3  Endometrial  ARID1A  SMARCD1_proteomics                True  1.744915e-04\n",
       "4  Endometrial  ARID1A    BCL7C_proteomics                True  4.379537e-04\n",
       "0  Endometrial  ARID1A   ARID1A_proteomics               False  1.025018e-10\n",
       "1  Endometrial  ARID1A    MAP1S_proteomics               False  1.010149e-07\n",
       "2  Endometrial  ARID1A     DPF2_proteomics               False  2.166609e-07\n",
       "3  Endometrial  ARID1A    KDM4B_proteomics               False  4.414547e-07\n",
       "4  Endometrial  ARID1A     RRM1_proteomics               False  4.455870e-07\n",
       "5  Endometrial  ARID1A   MAP7D3_proteomics               False  4.880890e-07\n",
       "6  Endometrial  ARID1A    CYTH3_proteomics               False  3.109396e-06\n",
       "7  Endometrial  ARID1A    CEP78_proteomics               False  3.749172e-06"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "if len(all_significant_comparisons) > 0:\n",
    "    display(all_significant_comparisons)\n",
    "    \n",
    "else:\n",
    "    print('No Significant Comparisons!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example analysis using DPF2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to create and trim down a dataframe in order to have our desired protein levels for each sample, along with a label as to whether it is a mutated or wildtype sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: In joining the somatic_mutation table, no mutations were found for the following samples, so they were filled with Wildtype_Tumor or Wildtype_Normal: 101 samples for the ARID1A gene (/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py, line 1)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>DPF2_proteomics</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>S001</td>\n",
       "      <td>-0.548</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S002</td>\n",
       "      <td>-0.919</td>\n",
       "      <td>Mutated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S003</td>\n",
       "      <td>-0.555</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S005</td>\n",
       "      <td>0.372</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S006</td>\n",
       "      <td>-0.642</td>\n",
       "      <td>Wildtype</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       DPF2_proteomics     Label\n",
       "Sample_ID                           \n",
       "S001                -0.548   Mutated\n",
       "S002                -0.919   Mutated\n",
       "S003                -0.555  Wildtype\n",
       "S005                 0.372  Wildtype\n",
       "S006                -0.642  Wildtype"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protdf = en.join_omics_to_mutations(omics_df_name='proteomics',omics_genes='DPF2',mutations_genes='ARID1A')\n",
    "protdf = protdf.loc[protdf['Sample_Status'] == 'Tumor']\n",
    "\n",
    "'''Create the binary valued column needed to do the comparison'''\n",
    "for ind, row in protdf.iterrows():\n",
    "    if row[gene+\"_Mutation_Status\"] != 'Wildtype_Tumor':\n",
    "        protdf.at[ind,'Label'] = 'Mutated'\n",
    "    else:\n",
    "        protdf.at[ind,'Label'] = 'Wildtype'\n",
    "\n",
    "'''Format the dataframe correctly'''\n",
    "protdf = protdf.drop(gene+\"_Mutation\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Location\",axis=1)\n",
    "protdf = protdf.drop(gene+\"_Mutation_Status\", axis=1)\n",
    "protdf = protdf.drop(\"Sample_Status\",axis=1)\n",
    "\n",
    "protdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will plot our data using the seaborn library as a stripplot on top of a boxplot in order to see the distributions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x122bf2208>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZUAAAEKCAYAAADaa8itAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAIABJREFUeJzt3Xl4lOXV+PHvmUkyWUlCEiDsUVBxQZRoVSy4IrWtSLVqLYpWpb6tuLS+dXtrq9W3al1+llpfrbWlda9WpSoiIKCtG8ENZF8EAoQlCdm3mTm/P2YyzCQTMhMmmSznc11czHM/9/PMiQ45c9/PvYiqYowxxsSCI94BGGOM6T0sqRhjjIkZSyrGGGNixpKKMcaYmLGkYowxJmYsqRhjjIkZSyrGGGNixpKKMcaYmLGkYowxJmYS4h1AV8vNzdWRI0fGOwxjjOlRli9fvldV89qr1+eSysiRIykqKop3GMYY06OIyJZI6ln3lzHGmJixpGKMMSZmLKkYY4yJmbgnFRGZIiJrRWSDiNwa5vwjIvK5/886EdkXdM4TdG5u10ZujDGmpbg+qBcRJ/AYcDZQDCwTkbmquqq5jqreFFR/FnBc0C3qVHVcV8VrjDHmwOLdUjkR2KCqm1S1EXgBmHqA+j8Anu+SyIwxxkQt3kllCLAt6LjYX9aKiIwACoB3g4qTRaRIRD4SkfPbehMRmemvV7Rnz55YxG2MMSaMeCcVCVPW1v7GlwAvq6onqGy4qhYClwL/T0QODXehqj6pqoWqWpiX1+7cHWOMMR0U78mPxcCwoOOhwI426l4C/DS4QFV3+P/eJCJL8D1v2Rj7MI0x3Vl5eTnPP/88W7ZsYfz48VxwwQU4nc54h9UnxTupLANGi0gBsB1f4ri0ZSURORzIBj4MKssGalW1QURygQnAA10StTGmW/ntb3/LmjVrAFi9ejWNjY1Mnz49zlH1TXFNKqrqFpHrgPmAE3haVb8SkbuBIlVtHib8A+AFVQ3uGhsDPCEiXnzdePcFjxozxnS92bNns2HDhi59T7fbzbZt20LKXnrpJV577TWGDAn7iLbLjBo1ilmzZsU1hq4W75YKqvoW8FaLsjtbHP86zHUfAMd0anDGmG7P4XDgcDjwer2BMlWlrq4ujlH1XXFPKsaY3iNe38qXLFnCY489RkNDA7m5ubhcLpKSknj00UfjEk9fZknFGNPjnXbaaZxwwgns2rWLESNG8LOf/SzeIfVZllSMMb1CWloahxxySLzD6PPiPU/FGGNML2JJxRhjgMbGRj7//HO2b98e71B6NOv+Msb0Og0NDTgckX9n3rFjB7fffjtlZWUAXHDBBcyYMaOzwuvVrKVijOk1amtr+cUvfsGOHTsoLi7mkUceIXR6W3gvvfRSIKEAvPrqq9g6gR1jScUY02ssWLAgMLMeYPHixaxYsaLd68rLy0OOvV4vFRUVMY+vL7CkYozpNcK1LiJpcZxxxhkhx8OHD+fQQ8OuT2vaYc9UjDG9xoQJE3jjjTcCs+tTU1MpLCxs97pJkyaRkJDA+++/z4ABA5g2bRoi4RZRN+2RSPobe5PCwkItKiqKdxjGmE6yfPlyfve73yEi3HvvvTZ3JUZEZLl/q5EDspaKMaZXGT9+PAMHDgSwhBIH9kzFGGNMzFhSMcYYEzOWVIwxxsSMJRVjjDExY0nFGGNMzMQ9qYjIFBFZKyIbROTWMOevEJE9IvK5/8/VQedmiMh6/x9bqMcYY+IsrkOKRcQJPAacDRQDy0Rkbpi95l9U1etaXNsf+BVQCCiw3H9tOcYYY+Ii3i2VE4ENqrpJVRuBF4CpEV57DrBAVcv8iWQBMKWT4jTGGBOBeCeVIcC2oONif1lLF4jIlyLysogMi/JaY4wxXSTeSSXc4jot1435FzBSVccCC4E5UVzrqygyU0SKRKTIlrM2xpjOE++kUgwMCzoeCuwIrqCqpara4D/8EzA+0muD7vGkqhaqamFeXl5MAjfGGNNavJPKMmC0iBSISBJwCTA3uIKI5Acdnges9r+eD0wWkWwRyQYm+8uMMcbESVxHf6mqW0Suw5cMnMDTqvqViNwNFKnqXOB6ETkPcANlwBX+a8tE5Df4EhPA3apa1upNjDHGdJm4r1Ksqm8Bb7UouzPo9W3AbW1c+zTwdKcGaIwxJmLx7v4yxhjTi1hSMcYYEzOWVIwxxsSMJRVjjDExY0nFGGNMzMR99JcxxsRCU1MTH3zwATU1NXg8HpxOZ7xD6pMsqRhjejyPx8Ptt9/O2rVrAXA4HAwePDjOUfVN1v1ljOnxvvjii0BCAfB6vVRWVsYxor7LkooxpsfzeDzxDsH4WVIxxvR4xx13HMOHDw8ciwgZGRlxjKjvsmcqxpgeLyEhgQceeIDFixdTXV3N0qVLSUxMjHdYfZIlFWNMr5Camsq3v/1tAD744IM4R9N3WfeXMcaYmLGkYowxJmYsqRhjjIkZSyrGGGNixh7UG2OM3yuvvMK7775LVlYW06dPZ8yYMfEOqceJe0tFRKaIyFoR2SAit4Y5/zMRWSUiX4rIIhEZEXTOIyKf+//MbXmtMcZEauHChcyZM4dt27axYsUK7rrrLmpqauIdVo8T15aKiDiBx4CzgWJgmYjMVdVVQdU+AwpVtVZE/gt4ALjYf65OVcd1adDGmC7R1NTE1q1befnll6murmby5Ml885vf7NC93G43r7zyCp999hkjR47k0ksvpV+/fiF1li9fHnJcW1vL6tWrKSws7PDP0BfFu/vrRGCDqm4CEJEXgKlAIKmo6uKg+h8B07s0QmNMl6quruaRRx5h2bJliAiqCvjW90pNTWX8+PFR3/PZZ5/llVdeAWDVqlVs27aNe++9N6TOiBEj+M9//hM4djgcDBs27CB+kr4p3t1fQ4BtQcfF/rK2XAXMCzpOFpEiEflIRM7vjACNMV3r2WefZdmyZQCBhNIs+Jd+sJUrV/LnP/+ZefPm0djY2Or8+++/H3K8YsWKVgtOTp06NdAqSUlJ4aqrrmLgwIEd/jn6qni3VCRMmYYpQ0SmA4XApKDi4aq6Q0QOAd4VkRWqujHMtTOBmUDI+kDGmO5n/fr1Bzy3a9eukF/2//73v3nggQcCxy27sZYtW8aePXtCyjIyMkhJSQkpS0lJ4c4776SiooLk5GRcLtfB/Bh9VrxbKsVAcPtyKLCjZSUROQu4AzhPVRuay1V1h//vTcAS4Lhwb6KqT6pqoaoW5uXlxS56Y0zMHXPMMW2e27JlC/fcc09I2Ztvvhly/Mknn9DU1BQ4fu6551q1eK6++uo21wbLzMy0hHIQ4p1UlgGjRaRARJKAS4CQUVwichzwBL6EsjuoPFtEXP7XucAEgp7FGGN6pksuuYTJkyeTkZHBYYcdRm5ubsj5LVu2sHu371eBqrbq7nI4HDgc+3+1tRzBJSKcfPLJnRS9iWtSUVU3cB0wH1gNvKSqX4nI3SJynr/a74B04B8thg6PAYpE5AtgMXBfi1FjphvxeDw0NDS0X9H0eS6Xi+uuu45nn32WBx98kMMOOyzkfHp6OllZWbjdbu64445W3WVTpkwJ2Up48uTJIedPO+00kpOTO+8H6OPi/UwFVX0LeKtF2Z1Br89q47oPgLbbyabbmDdvHn/729+oq6tj4sSJzJo1y5YlNxG74oor2LZtG9u2bSMtLY3rrruOpKQkFi5cyMqVK0PqnnfeeVx99dVcf/31iPge2V544YUMGDCAL774goKCAs4555x4/Bh9hrTsa+ztCgsLtaioKN5hHJTZs2ezYcOGeIfB9u3bARgypO0Be01NTRQXF4eU9e/fn8zMzJjFMWrUKGbNmhWz+5nuR1XZtWsX2dnZuFwuSktL+elPf0ptbW1IvUmTJrF27VpKSkpIS0vj6aefbvVA3nSMiCxX1XYn7cS9pWJ6rrq6unbrhBvead1gnaO7fNnoCuXl5a0SCsDSpUsDr2tqapg5cyb9+/fvytC6pa784mVJpQfqLt/Kb7jhBgAeffTRNuuUl5dz9dVXh4zGufrqqznzzDM7Pb6+ZsOGDaz/6jOGp/f+/drd3lSg/RZIXcVeGqo2dX5A3djWamf7lWIo4qQiIg8A9wB1wNvAscCNqvpMJ8VmeoHs7Gxuv/12/v73v1NVVcXZZ5/NGWecEe+weq3h6R5uP76y/Yo93O76Wh5YNYR6b/NYIyXctLfT82v47pDe/9/jQP73037tV4qhaFoqk1X1FyIyDd/8ku/jG3VlScUc0Pjx4wNLa9TW1vLss8+yadMmjj32WL7zne+EjNQxJhIDkt3cetR2PilNB+CtHVkh5wXlpNxqzsnfF4/w+rRokkrzcJ1zgedVtax5dIUxkXr44Yf55JNPACgqKqKiooLLL788zlGZnijX5ebcwb6kUdqQwMelGYFzlxXs5cSc6jav/XBvOm9uz6bBK3wjp5qRaQ3kpzQyJLWpzWtMZKJJKv8SkTX4ur9+IiJ5QH3nhGV6o9ra2sCaTs2WLl1qScUctEtH7uWIfnXsqEviqMxaRmW0PRikpC6R577ORf3dZUt27x+J+K38cr49xFo3ByPipKKqt4rI/UClqnpEpBbfisLG4PF4WLJkCZs2bWLcuHGccMIJreq4XC7S09OpqqoKlOXk5HRlmKaXcgqckFMDtL//yaZqVyChtLSgJIvTB1aSmuCNcYR9RzQP6n8KPKuqzUNLkoDvAX/sjMBMz/L444/zzjvvAPCvf/2La665hu9+97shdZxOJ0OHDmX16tWBsnPPPbdL4zQ9x5f7UplbnE2N28HJudV8Z0g5jgh73Bu9wvqqZLIS3a26tArSGxA0bGJxq9DoFVJj8QP0UdEs03KNqgbahapaDlwT+5BMT1NfX8+iRYtCylou8gewbdu2kIQCviXIjWnmVViwM5PffjWYP20YQEl9ElXuBN4pyeKjvekR3WNPfQJ3rRjK4+sH8dtVQ3lxS2hrOD+liUtGlJKR4CZBQlskR2XWkpXU+4dkd6Zonqk4RETUPwXfv2tjUueEZXoSp9NJYmIiHs/+f4zhVnkNtzWrbddqgr27qx+vbw8/WXF9VQqn5LX98L3ZwpJMKpr2/2p7f08/ThtYwcBkd6BsQl4VJ+ZU4xRlxb5UVlakMjC5kYl5VeFuaaIQTUtlPvCSiJwpImcAz+Obr2L6uMTERC688MLAscPh4OKLL25V77DDDqOgoCBwLCKcffbZXRKj6RlW7Etr89zwtMhWYqh2tx6iXt20v6zJC3M25fHzT0dw+xfDqfU4+OHIvZw1qJIkZ99atqozRNNSuQX4MfBf+GYZvQM81RlBmZ7noosu4rjjjmPjxo2MHTuWwYMHt6rjcDi45557ePPNNykrK2PixIkcffTRcYjWdFcDkpvYWB28grDiFDgxp5qJeZFNYvxGbjVfBCWngcmNFKTvT0jv7e7HsjJfV1q128nzX+dyeEY9/V3uVvcy0Ytm9JcXeNz/x5hWRo8ezejRow9YJyMjg0suuaSLIjI9zbmDy9lS42JHXRJJDi/fyi9nVEY9I9IaI35IPzarlv8aXUJRaTqZSW7OGFgRcu3W2tCuWS9CcW2SJZUYaTepiMhLqnqRiKwgzFa/qjq2UyIzxvQ52Ukebj9qO7vqE/iyPJU3dvTHo0Kuq4mfji4hLzmyX/xHZdZxVGb4BU9HZ9SzvGz/Q/9E8VKQblPuYiWSlsoN/r+/05mBGGNMs7QEL2/6EwrA3oZE3tqRzYxD9rRzZftOya2irCGBj0rTyUjwcN7QcjISbV5KrLSbVFR1p//vLQAi0i+S60zvVVxczK5du/B6vSHbthoTK5VNTtwa2t+1q/7Av3YavcK2miQGJjeRnuhlb0MCdR4HQ1MaCV5RyiFw3tByzhtaDviWbLl/1WCSHMo5+fs4so0WjolMNJMffwzcjW+ZluZuMAUOOZgARGQK8CjgBJ5S1ftanHcBfwPGA6XAxar6tf/cbcBVgAe4XlXnH0wspn1z5szhlVdeAXwP3vPz8+MckemN8pObGJTcSEn9/lkL22pdLNiZydn5Fa3qb6lJ4vH1g6h2O0kQL4ekN7C+KhlFGJ7awHWHlYSdJb+6Iplnv84LHD+5YSC/PLqYHHu+0mHRfM28GThKVUeqaoH/z8EmFCfwGPAt4EjgByJyZItqVwHlqjoKeAS433/tkcAlwFHAFOCP/vuZTrJ3715effXVwLHX62XfPlsnycSeCJw1cB/Bj3EV4Y0d2VQ1tf619Xpx/8BQYrc6WFeVEpgxv7XWxdLd4Zd/X1UROnfercKaStsp8mBE0421EWi91drBORHYoKqbAETkBXzria0KqjMV+LX/9cvAH8S3PPJU4AVVbQA2i8gG//0+jHGMxq+yshKvN/TbXvCERxNf27dvp6bK2eX7Z3SWXd4sWu6R4lHh4S8zSRRPi7oHnoc9b0cmi3ekkCF1OINm0ddo6++hC7YmsnRb7/hvCLClykmaf+vvrhBNUrkN+EBEPgYCg75V9fqDeP8hwLag42LgG23VUVW3iFQAOf7yj1pc2/Zm6eagFRQUUFBQwObNmwNlGRkZB7jCmI5RBU+YjpQE3CTQ+otMCg1UH2DFLi8OakmmURPIoyLwjCWVBhpIpN6/OEg69SSJdX0djGiSyhPAu8AKIFZDJcKNPG85bLmtOpFc67uByExgJsDw4cOjic8EERHuvvtuXnvtNXbv3s3atWtJS2t7BrTpWkOGDKHBvbPX7Pz4pw3JrSYxzjqsJOzaXF6t5GefjsCt+xNRgnhJdCh1nv2tETcJ/OjIOvJTfItMrq5IZsluD16tZdKASo7O6n1Di//30364hnTd9+1okopbVX8W4/cvBoYFHQ8FdrRRp1hEEoBMoCzCawFQ1SeBJwEKCwttHYYO8nq9bN26leOPP56jjjqKm266Kd4hmV7ssoI95O5oYkuNi1EZ9ZyTX0GiI/w/X4fA2YMqmLczO1D27cH7KK5LCpmT4sTLB3szyElyMyy1gcfXD8Lr/366viqFO47aHvFcGBNeNEllsf8b/78I7f4qO4j3XwaMFpECYDu+B++XtqgzF5iB71nJhcC7qqoiMhd4TkQeBgYDo4FPDiIWcwANDQ3ccccdrFu3DoAjjjjChhSbTpXsVKYNK29VXut2UON2tPrl/+0h+xiR1sDXNS4OTW9gTGYdu+sT+LraRWljIg4UL8LiXb5NuTIT3YGEAr4H/CsrUjk9uXe09OIlmqTS/Mv+tqCygxpS7H9Gch2+xSqdwNOq+pWI3A0Uqepc4M/A3/0P4svwJR789V7C91DfDfw0aK8XE2Pvv/9+IKEArFmzhtzcXHumYjrF3oYEMhI8uFos8Lh4Vz/mFmfTpA6GpzbwX6NLQiYuHp1Vx9FZ++eZDEh2c+cxxRTXJrGgJJPPy/e3WoJXMm6Wk2StlIMVzdpfBe3Xip6qvgW81aLszqDX9cD327j2XuDezojLhKqoaD03oOVIMGMOVkWjk//bMJBttS5cDi8XDCvlG7nVLC9LZ3N1Ev/e0y9kqPD8nVlcOPzAnSVOgRFpjaSEWYH4kLR6NtX4FrA8Pruao7NiPcC174lm8mMivhWKJ/qLlgBPqGpTmxeZXmPChAm8+OKL1Nf7HmSmpKRE9ZB++fLlvPXWWyQmJjJt2jQOP/zwzgrV9GBv7cxim3/Bxwavg5e25rC6MpnPysO3iHfXJ0Z879MHVvBZeRr1Hl+X7VGZtfzX6F3sqvd1jdmzlNiIpvvrcSCR/dsHX+YvuzrWQZnuZ9CgQdx///3MmzcPEeHcc8/l4Ycfjuja9evX85vf/CbQslm+fDmPP/44ubm5nRmy6YF21YUmCbc6+Ky87R0fx2ZH3rIYnNLEnUcX8+W+VPolejg603ftwGT7XhxL0SSVE1T12KDjd0Xki1gHZOLP6/Xy17/+lfnz55OSksJll13GmWeeSUFBAT/5yU+ivt8HH3wQ0lXW0NBAUVERU6ZMiWXYphc4JquWDdXtzWhXBriaODWvigm5ke3UuKfetyVxZZOTb+RUMda6uTpNNEnFIyKHqupGABE5BMLMQjI93uLFi3nttdcAqKur4/e//z1jxowJu/FWJAYMGBBRmTGnD6zk0/I0ttQEb9TVckqaMCKtkTMGRTZKq8kr/L+1+YEH819VpJIgu6Jq5ZjIRTMe9L/xDSteIiJL8U2E/HnnhGXiac2aNSHHqsratWs7fL8zzjiDsWP3b7szadIkxo0b1+H7md7LIZAbwWKO6YmRf5/dUJXcaqRXUZlN2u0s0Yz+WiQio4HD8X11WONfd8v0MmPGjGH+/P0LPjscjoN6sO5yubjnnnvYunUrCQkJDB48mNraWj7//HOysrI48siWa4iavuyU3Co+LUsLjPIamNzI8NQGlpX5HtZn+3dzjFR2mGHC4cpMbEQz+isV+BkwQlWvEZHRInK4qr7ReeGZrlBZWUlKSgqJib6HpKeffjpbtmxh/vz5pKamMn369A53fQVrXiJn586d3HLLLYEVjidOnMjNN9980Pc3vcPh/eq54fASlpWm0S/Rw8QBlWQkejljYCVVbiejM+pIjKKPZVBKE2cMrGDxLt9w5EHJjZwZYdeZiV40z1T+AiwHTvYfFwP/ACyp9FClpaU89NBDrFy5koyMDK655hpOO+00RIQrr7ySK6+8ssP33rhxI++99x5ZWVlMnjw5ZPjx66+/HrJk/nvvvccFF1xAQUGnTIUyPdCojHpGZYSuwzUsrbHD9/vesDImDqikusnJ8LSGsPvdL9nVj7d3ZuFROG1AJd8eYts6dEQ0SeVQVb1YRH4AoKp1/iXoTQ/T2NjI7NmzWbp0aaCsqqqKP/zhD4wfP/6gZ8mvWrWKO+64I7As/pIlS3jkkUcCS7pUV1e3uiZcmTGxlOtyt/m8ZktNEi9vywkcz9uZzfC0Bo7Jsl0goxVNUmkUkRT8KwGLyKEErQHWV8yePZsNGzbEO4yDsm/fPsrLW6+p1NjYyM0330xycnKYq0KpKhs3bgTghhtuCDm3e/fukH1WNm/ezLXXXktKim+oaF1d6D/UxMRE/vSnP9HTv6OMGjWKWbNmxTsM0wGbqlt/5jdVJ1tS6YBoksqvgbeBYSLyLDAB6Hj/SA+1YcMGPl+5Gk9q/3iH0mFpNOAK8/vbq/DVjn2E31VgP0HJoJ4Ep29J8TWbtlKDK3A+lUaSW9xi7c5yPOzvx07ARRIeFChvTGDX5t0d/XG6BWftwayrauLtkPTWS96HKzPti2b01zsishw4Cd9vnRtUdW+nRdaNeVL7U3fEufEOo8Nk3yZcuz4NHCvQmJJLxYBjaUrObvtCv4y9q0go3b85p0s8VA7/Bo0pvu6DpoZKkrYuxuH1zVSuS8uneuiE2P4Q3UzKmrfar2S6rRFpjXx/+F7m78zC7RVOH1hprZQOimb01yJVPRN4M0yZ6UFqMwtIaKoltfJrPM5kKvOOpiFtUMTXO92t/7E5gsrcrn7sOmQKydU78SQk05A6MCZxG9OZJg2oYtKAyGbom7a1m1REJBlIBXJFJJv9fSP98O1jYnoaESrzjqYy7+gOXV7bbxipFZsDHwSP00VDWmji8Dpd1GaOPLg4jTE9TiQtlR8DN+JLIJ8GlVcCj3VGUKZ7a0wdQOnQU0nbtxmvM4nq/oehjshXizWmu/lqXwrzdmbR5BUmDajklDwbjdhR7SYVVX0UeFREZqnq7C6IyfQADWmDouoyM6a72tuQwBMbB+JVX9v7uS15ZCd5GJNpz1Q6IprRX0+IyPXYfiomWqok1+zE2VhDfXo+nqS2lzI3B2drtZP//bRfvMOIG4868OCgrM63UfDA1PY3kqtRVyChNPvr+kwye0nre2u1k9Fd+H7RJJU/YvupmA7ILllGauVWAHTvCvYOPZXGVFulONZGjRoV7xDiKnj+lboUr9eLa0T7/020vp6KnTtDylJyh+HqJVtlj6ZrPxtx209FRPoDLwIjga+Bi1S1vEWdcfgSVz98y+zfq6ov+s/9FZgENK8sd4Wqft7ReEzncDbVBhIKgKiXjLJ1lFpSibm+PPGyoqIiZFkhEaFfv348+uijEV3/zDPP8Oqrr+LxeJg4cSLXX389CQnR/Ho0zeK5n8qtwCJVvU9EbvUf39KiTi1wuaquF5HBwHIRma+qzYvy/LeqvnwQMRgArweH1403wdV+3Whp633BUdvb3sRWeXk5bnfoEiwej4eamhqcTme7q0RMnz6d733ve3g8noNepqiviyapNO+nsgnfsOIRHNyM+qnAaf7Xc/A9owlJKqq6Luj1DhHZDeQBttLbQRKvG1dNCUl1ZaRVbMbhbaI+dQBlg09CnUkxex9PUhp16YNJqd4B+CZaVmd3ZQ+v6QtGjBjBiBEj2LJlS6BMVfnhD39IQkIC06ZN44c//CHgSzZr164lOzub/Pz8QP3U1NQuj7s3iiipiIgDqMPXPRer/VQGqupOAFXdKSIH7A8RkROBJGBjUPG9InInsAi4ta14RGQmMBP2L7/elzncdeRtWUyCO3Tnu+Ta3WSUrqZywLFtXNkxZYNPIqVyGwlN1dSnD45o1r4x0RARfv3rX/OPf/yDkpIS1q1bF1iktLGxkRdffJHx48eTm5vL7bffTklJCSLC+eeff1CrcZvWIkoqquoVkYdU9WTgy0hvLiILgXDjTu+I9B7+++QDfwdmqAb6Tm4DSvAlmifxtXLuDne9qj7pr0NhYWGY/pi+JW3fplYJpVliw/71uZJq95JUt5fG5P40ph3EMxBxUJc5ouPXGxOBnJwcrr32WsDXndXS119/zdKlSykpKQF8LZlXX32Vc845Jyb7BRmfaLq/3hGRC4B/qobrKG9NVc9q65yI7BKRfH8rJR8Iu6KgiPTDtzTM/6jqR0H3bh6u0SAifwFsl6cIOTxtjwJvnhmfVr6BrN37xz1U5h5FVc6YTo/NGPA9eE9KSgqsbB2tlJQUKiv3f0FyOp0ce+yxfPrpp63qlpWVWVKJoWiSys+ANHwP7OvwdYGpqnZ0UPxcYAZwn//v11tWEJEk4FXgb6r6jxbnmhOSAOcDKzsYR59TmzmCtIpNiL/R5xUH3oQUavsNDzxeWxVFAAAexklEQVTvSC8L3ZM+vWwtVf2PgB6+PL3p3hobG3nooYf46KOPSExM5KKLLuKiiy6K+j6pqank5OSQnp5OUlISF198Mfn5+Zx22ml89FHguykDBgzgiCOOiOWP0OdFs0pxrIdE3Ae8JCJXAVuB7wOISCFwrapeDVyEb7Jljohc4b+ueejwsyKShy+5fQ5cG+P4eq2m5Gz2DD+d1IoteJ0J1GQdijch9BthuNSRXLUDT2IKTSk9d9l/073NmzePDz/8EPAlmGeeeYYTTzyRkSNHRn2vcEOKTznlFG655RaWLFlC//79ueCCC2zocIxF9V9TRL4HnIpvEM/7qvpaR99YVUuBViscq2oR/gmVqvoM8Ewb15/R0fc2vsRScYAH5lX9DyNrd9A0JIWcnb5/7DX9RrAv/wQAXNU7Sav4Gq8ziar+h9tseXNQtm3b1qpsy5YtrZLK5s2bWbRoEampqUyZMoX+/SP/ojNhwgQmTOjdWzHEUzRL3/8RGAU87y+6VkTOVtWfdkpkJq5qskfT5MrCVbuHxLoyUmpLAufSKrdQ3f8wnO56crb/J9CqSa7eya5DpqAO++ZnOqbl41oRYezYsSFlmzdv5uabb6apyfdscNGiRfzxj3/E5eqEeVYmao4o6k4CzlHVv6jqX4Bz2T/PxPQW6iV13yaySpbjdNdTlXsk6mydJByeRlIqt4Z0kzk99bhqe/YOjia+gueZgC/JtNx+euHChYGEArBnzx6Kioq6JD7Tvmi+Uq4FhgPN/9eHEcXwYtMzZO36jLSKzQCkVWymoqma2syRpFQVBxJIU1IGjSm5JNfsanW9J6Fjo3WMAVp1YyUmJtKvX+hYoHCTFG3iYvcRTUslB1gtIktEZAmwCsgTkbkiMrdTojNdy+shteLrkKK0fZtoSBvE3qETqckcSWX/I9g7bBKIUJ09iqak/eM3ajJH2sRGc1AuvfRSMjMzAXA4HEyfPp309NDndN/61rfIzc0NHB9zzDEce2xsJ+yajoumpXJnp0VhugcR1JGAePd3LTRvvtWYNqDVBEhvQjK7R04mqa4UrzMJt6vvLrluYmPkyJE89dRTrF27lvz8fPLy8lrV6d+/P3/84x8pKioiNTWVcePG4XBE8/3YdKZohhQvPdB5EfnQP+Pe9FTioDL3qMCkR0WozD2qnWuExtTcA9cxJgoul6vVw/mWkpOTOfXUU7soIhONWA7TOfAyoKZHqMkeRUPqABLry2lMzcWTmBbvkEwfVVdX1+EZ9SZ+YplU+vyaWr2F29WvU7qykqu2k1axOTCnxe3KjPl7mJ6vuLiY3/3ud2zevJnhw4fz85//nIKCgniHZSJkHZGmS7hqdpGz40OSa0pIrdxK7ralyAHWIDN91x/+8Ac2b/aNQNy6dWvEG22Z7iGWScUWhTJtSqkKnSnt9DTiqm09JNmYjRs3hhxv2rQJr9c2duspYplULovhvUwv40loPY8gXJkxLR/SH3PMMTa6qwdp9/+UiAwTkRdE5H0RuV1EEoPOBdb+UlVbJdi0qTp7FI2urP3HmYfYwpQmrOuuu46TTz6ZzMxMTjjhBG688cZ4h2SiEMmD+qeBV4CPgKuApSLyXf+CkLbzUi/ncNfj8DbhTjq4RarVmcSeEWeS2LAPryPRFp40bcrOzua2226LdximgyJJKnmq+n/+17NEZDrwnoich4346tX67VlJetlaBKUhJYfSIafi8DTQr3Q1TncttRnDqc2KYlSOiM24Nx2yfPly3nzzTZxOJ9OmTePII4+Md0imDZEklUQRSVbVevAtRy8iJcB8fJt29Snbt2/HWVtBypq34h1Kp3LiJUPqA8euulKy1r+JCw9O8X2XcNXuIbFkBY0xHZneMzlrS9m+3R3vMHqljRs38pvf/CbwsP6zzz7jscceY+DAgXGOzIQTydOvp4BvBBeo6kJ8m2rZc5ReykHr0TYJeAMJpVkSHftF6sRLEm7EGrumHR9++GHI6K/GxkaWLVsWx4jMgbT7FVNVH2mj/DPg7JhH1M0NGTKEkoYE6o44N96hxFxifTn99n6Fw11PbcYheMvW4ghaB6xm0DgSd30aMna8MbuAuoHHRfU+GXu/ol/pagC84qR06Dd7/FIvKWveYsgQ++bcGQYMGNCqrOWwY9N9RDL6652g1zF7eiYi/UVkgYis9/8dtrNdRDwi8rn/z9yg8gIR+dh//Yv+/exNB4nXTe6290muKSGpYR9Ze1dSnT2auvTBNKTkUj6okNqsQ6jMPQr1p5WmpH5U5ezf3zupbi8pldsQT2PQfZtI3beZ1H0bEU8j4mkko2xt4LxDPWSUruq6H9T0OKeffjrZ2aG/HpYuXUplZWWcIjIHEtGD+qDX3wd+G6P3vhVYpKr3icit/uNbwtSrU9VxYcrvBx5R1RdE5P/wjUx7PEax9TlJdXtxeBtDyhIbKigbckpIWXXOGOrT8hFVmpKzQHwJJqukiDT/svleRyJ7hp+GJzGNvK8XkdhUDUC/0tXsHXwqoqFda8GtIWNaSkxMJD8/n/Ly8kCZ2+2mvLy81V4rJv4ieabSWZ3eU4E5/tdzgPMjvVBEBDgDeLkj15vW3EkZrf5HtxxG7GysYsDmdxi4ZSE5O/5DUu2eQHla0D4sDm8TGWXrSKkqDiQUAKe7nuS6XdSnDQq5b02mretkDmzixIkhxyNHjmT48OFxisYcSCQtlUP83U4S9DpAVc/r4HsPVNWd/nvsFJHWHac+ySJSBLiB+1T1NXwbhu1T1eanxMXAkLbeSERmAjMB+yC2wZOYRmXeWPrtXYmol4aUXKr6HxZSJ2v3FyQ2+rocnO56skuWseuQc8O2NMTbSNjvI6qUDT6JtH2bSGispD4tn/qMNv/XGQPAueeeS0JCAh9++CH5+flceOGFiNjKUN1RJEllatDrB6O5uYgsBAaFOXVHFLcZrqo7ROQQ4F0RWQGE60xts0Wlqk8CTwIUFhbacKM2VPc/jJrMAhzeJjyJQUuoeD0k15SQWF8WUj/BXYd4m2hyZdPoyiapwdc9oUBtZgENKXlkJK4moakWAI/TRW3mCNSRQHWLhGVMeyZPnszkyZPjHYZpRySjv5aKyHHAocBXqro60pur6lltnRORXSKS72+l5AO727jHDv/fm/zbGB+Hb4Z/logk+FsrQ4EdkcZl2qbORDyO/R8L8TSSt3UxiY1Vreo2JmejTt/4iL3Dvknavo04m+qo6zeUxlRfw3PPiDNJqdyKqJfafsPx2h72xvRqkYz+uhN4EbgAeFNEronRe88FZvhfzwBeD/Pe2SLi8r/OBSYAq1RVgcXAhQe63kQnsa6MAZvfYci6V8jZ9h4Odz2plVtaJRSP00Vd+hDKBp8UKFNnEtU5Y6gYdHwgoQB4nS5qskdT3f9wSyimU1VVVfHZZ5+FPNA3XS+S7q+LgXGqWisiOcDbwJ9i8N73AS+JyFXAVnwjyxCRQuBaVb0aGAM8ISJefAnwPlVtHn96C/CCiNwDfAb8OQYx9V2q9N/5MQlNNQAk1+4mc/fnYTfSqsw9ktqsQ7s6QmPatHz5cu677z4aGhpISEggOzub9HRbXy4eIkkq9apaC6CqpSISkzWo/QtSnhmmvAi42v/6A+CYNq7fBJwYi1gMODwNgYTSLKm+jMrcY0gvW4vD6xsT4XG6qE8fGo8QjWnTX/7yFxoaGgDfcOOysjLS0vrcKlLdQiRJ5dCgEV/S4vhgRn+ZbsTrdNGUlBHS1dWYkocnKY09I84kdd9mEAc1WQV4E1xxjNSY1lp2eXk8njhFYqId/QVRjgAzPYQIZYNPIqvkUxIb9tGQNoiKAb7NktxJGVQOGNvODYyJn9NPP525c/fPdkhPT7chx3ES0eiv5tcikucv29OZQZn4cLsy2Tvi9JjfN6ViC8k1O3En9aM6ezTqTGz/ImOicOWVV5KXl8fKlSsZPXo077//frxD6rPaTSr+2et3ArPwdX85RMQNzFbVuzs5PtMFEuvKAr/06zKGBpZeiYW08vVk7f4icJxUt5fSYRMPcIUx0XM6nUydOpWpU30dK//5z3/iHFHfFclD9xuBU4ETVDVHVbPxLYU/QURu6tToTKdLriomb+u79CtdTf+dH5NdUhTT+6dWbAl9v9rdOJrqYvoexpjuI5KkcjnwA1Xd3FzgH3k13X/O9GAZZetClrJPqdxCyr6vQVvvp9IRXmfoQ32vOFGHbeplTG8V0c6Pqrq3ZaGq7hGRPtk57qwt6zU7Pwr1BGcVAfrvKqKp5FOqcBFysgVHvW+1HG9y2yvFNuAhCXAIqEK9OkhevyA2wXcjztoywPZT6c4aGhrYs2cPgwcPxuGIycwIE0YkSaWxg+d6pVGjRsU7hJiqra1l165drcoTxcuRgzJJSWl7FvyGDb7hx6MOOfAvU6/XS319PYmJiSQm9tbvIQN73WejN/noo4949NFHqampYdCgQfzyl79k2LBh8Q6rVxLfiicHqCDiAWrCnQKSVbVH/ZYoLCzUoqLYPjfo6bZu3crs2bNZu3ZtSPnFF1/MoYceyvjx48MmgxtuuAGARx99tEviNN3f7Nmz2bBhQ7zDCMQwatQoVJWtW7eGbEmckpLCoEHh1rqNrVGjRjFr1qxOf5+uICLLVbWwvXqRDCl2xiYk010NHz6ca6+9ll/84hc0NfmWsU9MTOTFF18EYMSIETzwwAMHbLUY050Ef1a9Xm9IQgECn3MTe/bE1ABw6KGH8sgjj7B06VK2b9/OBx98EDi3ZcsW3n//fVt23LSru34rv/nmm1m3bl3g+Nvf/jZXXXVVHCPqvexplQkYPnw4l112GYcffnirc3V1NgzY9Fy33XYbZ5xxBoceeigXXnghl19uA1c7i7VUTCsTJ07kH//4B9XVvq2AMzIyOPXUU+MclTEdl5OTw4033hjvMPoESyqmlZycHB566CHeeecdAM455xxycnJa1XO73ZSWlnLNNdcwduxYfvSjH9nKsMb0cZZUTFj5+fnMmDHjgHV2795NQ0MDtbW1LFiwALfbzU032SILxvRlllRMh9TX1wf2r2j26aefhq3773//mw8//JBBgwZx/vnnk5GR0RUhGmPiwJKK6RCXy0VCQgJutztQNnLkyFb1FixYwOzZswPHX3zxBQ8+aLsnGNNbxW30l4j0F5EFIrLe/3d2mDqni8jnQX/qReR8/7m/isjmoHPjuv6n6LtEhLy8PJxO3zSmYcOG8eMf/7hVvUWLFoUcr1u3juLi4i6J0RjT9eLZUrkVWKSq94nIrf7jW4IrqOpiYBz4khCwAXgnqMp/q+rLXRSvaSE5OZlhw4Zx1113kZ3d6jsBAFlZWSHHTqfT9g43pheL5zyVqcAc/+s5wPnt1L8QmKeqtZ0alYmKiLSZUMC31EvwM5Tvf//7rRKNMab3iGdLZaCq7gRQ1Z0iMqCd+pcAD7cou1dE7gQWAbeqakPry0w8FRQU8NRTT/HVV18xaNAghg4dGu+QjDGdqFOTiogsBMKt2nZHlPfJB44B5gcV3waUAEnAk/i6zsLuRCkiM4GZ4Js1brrWnj17WLhwIaWlpUyaNInvfOc78Q7JGNNJOjWpqOpZbZ0TkV0iku9vpeQDuw9wq4uAV1U1sApccysHaBCRvwA3HyCOJ/ElHgoLCw+8LLOJqcbGRn75y19SXl4OwNq1a0lKSrJ1xIzppeL5TGUu0Dy7bgbw+gHq/gB4PrjAn4gQEcH3PGZlJ8RoDtK6desCCaXZxx9/HKdojDGdLZ5J5T7gbBFZD5ztP0ZECkXkqeZKIjISGAYsbXH9syKyAlgB5AL3dEHMJkqDBg1qtcve4MGD4xSNMaazxe1BvaqWAmeGKS8Crg46/hoYEqbeGZ0Znzk4Xq+XrVu3kpuby4wZM3jmmWdoampi1KhRXHjhhfEOzxjTSWxGvYm5kpIS7rrrLrZv305SUhJXX301f/3rX6msrGTIkFbfD4yJudLSUu666y5+9atfhV0M1XQe20/FxNxzzz3H9u3bAd+D+qeeegoRsYRiOqykpIT777+fG264gWeffRaPx3PA+nPmzGHFihX87W9/66IITTNrqZiYKykpCTlubGykrKzMZtKbDlFV7r777sDyPps3b0ZEuPTSS8PWLy0t5e2330ZVefvtt7n88suttdKFrKViYu7kk08OOR46dCjDhg2LUzSmp9uxY0er9eKWLVvWZv05c+YE9qT3eDzWWuli1lIxMTd16lQAPvzwQ/Lz87n00kvxjfw2Jno5OTmkpKSEbGl9oJUZFi5cGFg92+12s2DBAtvnpwtZS8XEnMPhYNq0aTzwwAPcdNNNDBw4MN4hmR4sOTmZn/zkJ6SmpgIwYsQILrvssjbrn3XWWSQk+L4vJyQkcPbZZ3dJnMbHWirGmG5v0qRJnHTSSZSXlzNoULiVn/abMWMGb7/9NuBbFfvyyy/vihCNn7VUjDE9gsvlajehgK+7bMqUKYgIU6ZMsYf0XcxaKsaYXmfGjBl8/fXX1kqJA0sqhm3btvGHP/yBTZs2ceyxx3Ldddd12p4npaWlpKWlkZyc3Cn3NwZ8rZXf//738Q6jT7LuL8MDDzzA6tWraWho4JNPPuGJJ56I6vqamhqWLFlCUVFRm5PSqqurueOOO7jyyiu5/PLLeeONN2IRujGmm7GWSh9XU1PDli1bQspWrVoV8fVNTU385Cc/CaxEfOyxx3L33Xe3GkL8z3/+kxUrVgBQX1/PU089xUknnURubu5B/gTGmO7EWip9XFpaWquJiYcddljE11dVVYUsbf/FF1+wcmXrXQi2bdsWcuz1eltNaDPG9HyWVAw///nPKSgoAGDs2LFce+21EV/bPHM5WH19fauy8ePHhxxnZGRwxBFHRBmpMaa7s+4vwyGHHMKjjz6KqkY98z0jI4P6+nqamnybcg4ePJhx48a1qnfOOecEnr3079+f6dOn28N6Y3ohSyomoCNLqbhcLh588EGWLFlCWloaU6ZMITExMey9L7jgAi644IJYhGqM6aYsqZiDVlBQEOg+M8b0bXF7piIi3xeRr0TEKyKFB6g3RUTWisgGEbk1qLxARD4WkfUi8qKIJHVN5MYYY9oSzwf1K4HvAe+1VUFEnMBjwLeAI4EfiMiR/tP3A4+o6migHLiqc8M1xhjTnrglFVVdrapr26l2IrBBVTepaiPwAjBVfJ3/ZwAv++vNAc7vvGiNMT1JaWkp119/PaWlpfEOpc/p7kOKhwDBExyK/WU5wD5VdbcoN8b0MnV1dbz55ps888wzrSbqtsW2E46fTk0qIrJQRFaG+TM10luEKdMDlLcVx0wRKRKRoj179kT41saYePN6vfzP//wPTzzxBC+99BI33XQTa9asOeA1LbcTttZK1+rUpKKqZ6nq0WH+vB7hLYqB4OneQ4EdwF4gS0QSWpS3FceTqlqoqoV5eXkd+VHMQSouLmbu3Ll89tlnqLaZ/40JsWbNGtavXx84drvdzJs374DX2HbC8dXdu7+WAaP9I72SgEuAuer7rbQYuNBfbwYQaaIyXayoqIhZs2bx1FNP8atf/Yonn3wy3iGZHqJ5B8dg4eZBBQu3nbDpOvEcUjxNRIqBk4E3RWS+v3ywiLwF4H9mch0wH1gNvKSqX/lvcQvwMxHZgO8Zy5+7+mcwkXn55ZdDVi+eN28elZWVcYzI9BSHHXYYxx9/fOA4NTWV7373uwe8xrYTjq+4TX5U1VeBV8OU7wDODTp+C3grTL1N+EaHmW6u+VtjM1UNu2aYMeH88pe/ZNmyZZSXl/ONb3yD/v37H7C+bSccX929+8v0Auedd17IEjATJ07stE3ATO/jdDo56aST+Na3vtVuQgHbTjjebJkW0+kmTpxIXl4eRUVFDBs2jG9+85vxDsn0cradcPxYUjFdYsyYMYwZMybeYZg+wrYTjh/r/jLGGBMzllSMMcbEjCUV0yabpGiMiZY9UzGtvPvuu8yZM4fq6mrOOussZs6cidPpjHdYxpgeQPrat9HCwkItKiqKdxgHZfbs2WzYsKFT7t3U1ERxcXFIWU5ODv369WtVtzmGUaNGdUoskRo1ahSzZs2KawzG9HYislxV29z7qpm1VEyIxsbGVmUNDQ1h66akpHR2OMaYHsaSSg/Umd/K9+zZw8yZM0OWVbniiiuYPHlyp72nMab3sAf1JkReXh4333wz+fn5pKenM23aNM4666x4h2WM6SGspWJamTBhAhMmTIh3GMaYHshaKsYYY2LGkooxxpiYsaRijDEmZiypGGOMiRlLKsYYY2ImntsJf19EvhIRr4iEnaUpIsNEZLGIrPbXvSHo3K9FZLuIfO7/c264exhjjOk68RxSvBL4HvDEAeq4gZ+r6qcikgEsF5EFqrrKf/4RVX2wswM1xhgTmXjuUb8aCNlmNkydncBO/+sqEVkNDAFWtXmRMcaYuOkxz1REZCRwHPBxUPF1IvKliDwtItlxCcwYY0xAp7ZURGQhMCjMqTtU9fUo7pMOvALcqKqV/uLHgd8A6v/7IeBHbVw/E5jpP6wWkbWRvrdpVy6wN95BGBOGfTZja0QkleK+9L2ILAFuVtWw69GLSCLwBjBfVR9uo85I4A1VPbqTwjRtEJGiSJbDNqar2WczPrp195f4Hrj8GVjdMqGISH7Q4TR8D/6NMcbEUTyHFE8TkWLgZOBNEZnvLx8sIm/5q00ALgPOCDN0+AERWSEiXwKnAzd19c9gjDEmVNy7v0zPJiIzVfXJeMdhTEv22YwPSyrGGGNipls/UzHGGNOzWFLpY0REReTvQccJIrJHRN5o57osEflJBPePqF6Y634tIjdHe53p+UTkERG5Meh4vog8FXT8kIjcLiIv+49Pa+vzKiJfi0huRz+H5uBZUul7aoCjRSTFf3w2sD2C67KASP6RRlrPmGYfAKcAiIgD3/ySo4LOnwIsUtULo7infQ7jxJJK3zQP+Lb/9Q+A55tPtGwxiMhK/zyg+4BD/SPwfici6SKySEQ+9Y/Cm+q/JKSe/x7/LSLL/Ksf3BV07ztEZK1/kuzhnfkDm27tP/iTCr5kshKoEpFsEXEBY4ByEWk1bUBEckTkHRH5TESeAJrXfWr5ef170GcUEXlWRM4TkStE5HURedv/WfxVUJ3pIvKJ/x5PiIizk37+XsWSSt/0AnCJiCQDYwld+qYttwIbVXWcqv43UA9MU9Xj8Q3pfsg/ryiknohMBkYDJwLjgPEiMlFExgOX4Ft653vACTH+GU0Poao7ALeIDMeXXD7E95k8GSgEvgQa27j8V8C/VfU4YC4w3F/e8vP6FHAlgIhk+t+neerCicAP8X0+vy8ihSIyBrgYmKCq4wCPv45pRzxXKTZxoqpf+lsfP2D/P6xoCfC/IjIR8OJb6HNgmHqT/X8+8x+n40syGcCrqloLICJzOxiH6R2aWyunAA/j+zydAlTg6x5ry0R8X0pQ1TdFpDxcJVVdKiKPicgAf/1XVNXtX9B2gaqWAojIP4FT8a2QPh5Y5q+TAuw+2B+yL7Ck0nfNBR4ETgNygsrdhLZgk9u4/odAHjBeVZtE5Os26grwW1UN2eLA/2DWxrObZs3PVY7B1/21Dfg5UAk83c61kX6O/o7vc3sJoesEtrxe8X1u56jqbRHe2/hZ91ff9TRwt6quaFH+NXA8gIgcDxT4y6vwtS6aZQK7/QnldPYvNtey3nzgR/5FQRGRIf5vi+8B00QkRXx75Xw3Zj+Z6Yn+A3wHKFNVj6qW4XvYfjK+7rC2vIe/W0pEvgU0r1be8nMI8FfgRgBV/Sqo/GwR6e8fvHK+P5ZFwIX+zyr+8xEtqNjXWUulj1LVYuDRMKdeAS4Xkc+BZcA6f/1SEfmP/2HpPOB+4F8iUgR8DqwJV8//XGUM8KG/G6EamO7feO1F/7VbgPc78+c13d4KfKO+nmtRlq6qe5u/lIRxF/C8iHwKLAW2QvjPoaruEt+eTK+1uMe/8bViRgHPNS9uKyL/A7zjH5HWBPwU32fVHIDNqDfG9AkikoovUR2vqhX+siuAQlW9Lp6x9SbW/WWM6fVE5Cx8renZzQnFdA5rqRhjjIkZa6kYY4yJGUsqxhhjYsaSijHGmJixpGJMJxGR6ijqRr1KczT3N6arWFIxxhgTM5ZUjOlCIvJdEfnYv6ruQhEJXi/tWBF5V0TWi8g1QdeEXeXZmO7IkooxXevfwEn+VXVfAH4RdG4svi0JTgbuFJHBba3y3MUxGxMxW6bFmK41FHhRRPKBJGBz0LnXVbUOqBORxfgSyamEX+X5va4L2ZjIWVIxpmvNBh5W1bkichrw66Bzba2W22qVZ2O6K+v+MqZrZbJ/++YZLc5NFZFkEcnBtyXBMtpe5dmYbslaKsZ0nlQRKQ46fhhfy+QfIrId+Ij9WwsAfAK8iW/3wt/4d0TcEW6VZ2zDKNNN2dpfxhhjYsa6v4wxxsSMJRVjjDExY0nFGGNMzFhSMcYYEzOWVIwxxsSMJRVjjDExY0nFGGNMzFhSMcYYEzP/H/UGQiM3ZMHdAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.boxplot(data = protdf, x = 'Label', y = 'DPF2_proteomics')\n",
    "sns.stripplot(data = protdf, x = 'Label', y = 'DPF2_proteomics', jitter = True, dodge = True, color = '.3')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
